KEP-2837: Introducing Pod level resource specifications with criteria for alpha phase (#4678)
kubernetes/enhancements: keps/sig-node/2837-pod-level-resource-spec/README.md

lacks a mechanism for sharing resources between containers easily. 

Introducing pod-level resource requests and limits offers a more simplified resource management for multi-container pods as it is easier to gauge the collective
resource usage of all containers in a pod rather than predicting each container’s individual needs. This proposal aims to support pod-level resource requests 
and limits that would allow Kubernetes to manage the total resource consumption of the pod, freeing users from the burden of meticulously allocating resources
for each container. This simplified approach is particularly valuable for tightly coupled applications, and it will enable Kubernetes to leverage the underlying
cgroup resource management mechanisms more effectively for better resource utilization for bursty workloads. 

This new feature of resource specification at pod-level would complement, not replace, the existing container-level limits, providing users with greater flexibility.

### Goals

This proposal aims to:

1. Extend Pod API to allow specifying resource limits on pod level in addition to individual container level resource specifications. 

> 2. Support Pod-level resource requests/limits as an opt-in feature that is compatible with existing usage of container level resource/limits.

Reviewer: what's the difference between these two items? For me they read almost the same

Author: Fixed the sentence

> 

Reviewer 2: is VPA in scope? how about in-place VPA?

Reviewer 2: nope: https://github.com/kubernetes/enhancements/pull/4678/files#diff-92bceaa7973d99b42530dc6380537c3d21f98ec8517e7b431bb6f3fcffa96398R623

Without this feature, the developer would need to decide a-priori how much resources to allocate to the tools - and this is not easy to do for this workload.

#### Story 2

Consider a scenario where a pod has some containers with occasional resource spikes while other containers often have idle resources. With current container 
level resource settings, the peak usage of each container needs to be accounted for when setting limits. This leads to wasted resources when the spikes are 
not occurring. Pod-level resource limits will allow setting a limit for the entire pod, enabling containers to share resources more efficiently without the
need to over-provision for each container’s peak usage. 

### Notes/Constraints/Caveats

1. Community is trying to move cgroupv1 to maintenance mode in future Kubernetes versions (likely 1.31). Hence this feature will be supported only on cgroupv2.
2. This feature is an **opt-in** feature, and will have no effect on existing deployments. Only deployments that explicitly require this functionality should
turn it on by specifying the relevant resource section at pod-level in the Pod specification.

> ### Risks and Mitigations

Reviewer: I think one of the big risks would be numerous custom schedulers and web hooks that might have a lot of assumptions about how the requests and limits are being calculated. Some of them may break

> ### Risks and Mitigations

Reviewer: another risk is 3rd party tooling taking dependency on the cgroup structure.

Reviewer 3: Could you provide an example? This doesn't change the cgroup structure, but rather what values set in the pod level cgroup. For monitoring tools, the change shouldn't have any impact IIUC.

Reviewer 4: @yujuhong that's a great callout - I was not sure if that was the case.  Let's make sure to highlight - the per-pod cgroup already exists, it is not a new level of nesting.

Reviewer 5: This will affect introspection scenarios. We bind mount container's own cgroups as read-only inside the container and those values are used by language runtimes such as JVM to tune itself. Downward API may be a mechanism to provide the pod level values.

Reviewer: what @mrunalp said is definitely something that may affect apps.

I meant any tools that took dependency on the cgroup limits and, for example, adjusting those limits if not set. Those may not expect higher level limits set

Author: I have updated the risks and mitigations, and the table & critical scenarios also talk about limits being set only in cgroup values for containers for runtimes relying on these values.

> ### Risks and Mitigations

Reviewer: another risk from the comment below is potential abuse of an issue. For example, if I have many low traffic web servers, I will likely make them best effort or bursteable. WIth Pod level requests and limits, it will be now relatively  easy to pack 20 of such containers into a single Pod and make it Guaranteed.

Yes, customer will start experience a new failure mode - noisy neighbours inside the Pod where a single container may bring the whole Pod down by exceeding limits. But the same time, there are benefits from running bigger Pods.

Exploring this abuse vector would be interesting for this KEP

Reviewer 3: I think we can call that out in the KEP to warn users. Users should set individual container limits if they suspect that the container could potentially have high memory usage.

We can also talk about the difference in OOM behavior as well.

Reviewer 4: OOM behavior seems like a critical aspect.  I am less worried about people over-packing a pod -- allowing users to opt-in to trouble is a core competence of Kubernetes.

And people do this already by forking too many nginx servers or whatever.  It's not really new.

Reviewer: The fact that there is now a pod level limit that can be set can motivate customers to abuse it this way. As a mitigation, will this KEP be the right place to add the limit on number of containers per pod?

Reviewer 3: > will this KEP be the right place to add the limit on number of containers per pod?

IMHO, this is not the right place to do it. As Tim mentioned, overpacking is not a new issue and instead of setting a hard limit, cluster admins can set up an admission plugin to limit this based on their own needs.

Author: Resolving this. Limiting the number of containers in a pod needs to be addressed in a separate KEP.

Reviewer 6: Agree with the over-packing a pod is NOT an urgent issue to consider.

> Since this is an **opt-in** feature, there should be no risk in merging the changes. Existing workloads that do not use this feature won’t be affected.

Reviewer: > Existing workloads that do not use this feature won’t be affected.

Theoretically they can if we will apply bad limits to the Pods which opted-in for the feature and those bad limits create a noisy neighbour situation for existing workloads

Reviewer: this is definitely something to cover in rollout section of PRR

Author: I added a sentence to address this. I feel the problem of bad limits to new pods creating a noisy neighbour is not limited to this feature. It could also happen is bad container level limits are applied.

Besides, the feature will be introduced in a phased approach, starting with an alpha release for testing and gathering feedback, followed by beta and GA releases
as the feature matures and potential issues are addressed.

> ## Design Details

Reviewer: can you please add the implementation details on how cgroup structure will be changed and whether any changes will be needed on container runtime side?

Reviewer 3: Here's my understanding:

The cgroup structure shouldn't be changed. We already have pod-level cgroup today. What would change is the value. 

For the container runtime, we already pass the parent cgroup (which is pod level cgroup) anyway, so there shouldn't be any change. 

As for VM runtimes, I mentioned this in another comment that we can add a brief section to talk about it.

Reviewer 6: +1 on what Yuju said. No changes on cgroup structure from the top to the container level, and the only changes are the value for the pod-level.

Reviewer 4: What I am mostly looking for in this KEP is information about how resources behave inside a pod:

* If pod > sum(containers), do they get equal access or first-come, first-served or ...
* If pod < sum(containers), how is OOM handled
* How does this affect requests vs limits

and so forth

  ```yaml
  apiVersion: v1
  kind: Pod
  ...
  spec:
    containers:
    - name: container-1
      resources:
        requests:
          memory: 100Mi
          cpu: 1
    - name: container-2
  ...
  ```

  No changes required to support this.

* **Scenario 2: Only pod-level resources are specified**

  ```yaml
  apiVersion: v1
  kind: Pod
  spec:
    resources:
      requests:
        memory: 100Mi
        cpu: 1
  ...
    containers:
    - name: container-1
    ...
    - name: container-2
  ...
  ```

  Following validation logic needs to be added for `resources` specified in the PodSpec:

  1. Only CPU and memory resources are supported in pod level resource specifications, hence invalidate pod-level resource specification for other resources 
  like gpu, storage, etc.

* **Scenario 3: Both pod-level and container level resources are specified.**

  ```yaml
  apiVersion: v1
  kind: Pod
  ...
  spec:
  resources:
    requests:
      memory: 100Mi
      cpu: 1
    containers:
    - name: container-1
        resources:
          requests:
            memory: 100Mi
            cpu: 1
    ...
    - name: container-2
  ...
  ```

>   Following validation logic needs to be added:

Reviewer: please also add the validation of requests comparing to limits. Can limit defined on the pod be lower than sum of requests of containers?

Author: Added the validation rules. However for specifically pod level limit to be lower than sum of requests of containers, that is ensure by the other validation rules mentioned in this section.

                  limits:
                    memory: 200Mi
                    cpu: 1 
          ```

          The pod-level limit will serve as an aggregate ceiling. This will cover the scenario where the sum of peak usage for individual containers in a pod is high, but these
          peaks don’t necessarily occur simultaneously.

          If we simply set the pod-level limit as the sum of the individual container limits, we might overallocate resources. The pod might never actually use all those resources
          at once, leading to wasted capacity on the node.

      * `pod.spec.resources.limits[memory/cpu] < pod.spec.containers[i].resources.limits[memory/cpu]` is invalid.

        A single container’s limit should not be allowed to exceed the pod level limit.

#### Scheduler Changes

The scheduler logic needs to be updated to support pod-level requests and limits.

* **Scenario 1: Only container-level resources are specified**

  No changes, same as existing.

* **Scenario 2: Only pod-level resources are specified**

  Consider pod-level requests to find the suitable node for pod placement.

* **Scenario 3: Both container-level and pod-level resources are specified**

>   Scheduler will prioritize pod-level requests as the sole indicator of the pod’s resources needs. It will ignore container level requests. This ensures that the pod 

Author: Agreed, reframed the sentence.

  is scheduled on a node that can fulfill the overall resource requirements of the pod, regardless of the individual container requests.

#### QoS Changes

The QoS definitions need to be updated to consider pod-level resources. To attach a QoS class to a pod, the  pod’s resource specifications are evaluated in order:

1. The values for resources specified in the pod’s `resources` stanza.
2. The value of resources specified in the container’s `resources` stanza.

> In simple words, if resources are specified at pod-level, the values in pod-level `resources stanza` are given a preference in determining the QoS class.

Reviewer: if pod limit is not set, will the sum of container limits be used instead? Or pod requests and limits must be set so the pod can become guaranteed?

Reviewer 6: It depends on QoS. For guaranteed, pod limit should be something like:

Pod_limit = max(init container's limit, sum of app containers' limit) + sidecar container limit + pod overhead

Author: If pod level `resources` is set, the container-level `resources` won't be considered. I reframed this section to add more clarity. PTAL

* **Guaranteed QoS Criteria**

    One of the following two conditions must apply:

    * Pod level requests and limits are specified for each resource, and request is equal to limit for each resource. The container level resource specification 
    (specified or not specified) is not taken into account for QoS class determination in this case.

>     * In absence of pod-level resources specifications, every container in the pod has requests and limits specified for each resource and all container requests

Reviewer: in absence - does it mean both needs to be omited? If pod has limit defined, but not requests, than it is NOT guaranteed, correct?

Author: Correct, Reframed this section.

     and limits are equal.

* **Burstable QoS Criteria**

    For Burstable QoS, the pod does not meet the criteria for Guaranteed QoS class and one of the following two conditions must apply:

    * Pod level requests or limits is specified. If both pod-level request and limit are specified, the request should be less than the limit. The container level
    resource specification (specified or not specified) is not taken into account for QoS class determination in this case.
    * In absence of pod-level resources specifications, at least one container within the pod has requests or limits specified. If both request and limit are 
    specified for a resource at container-level, the request should be less than the limit.

* **BestEffort QoS Criteria**

    None of pod-level or container-level resources requests or limits are specified.

#### Init Containers

> With the current implementation (before this KEP), Pod’s effective resource request/limit is the higher of:

Reviewer: please include the sidecars in this formulae. The calculation is more complicated these days

Reviewer 3: +1 for expanding this to include sidecar containers

  kind: Pod
  spec:
    resources:
      requests:
        memory: 100Mi
        cpu: 1
      limits:
        memory: 200Mi
        cpu: 2
    containers:
      - name: container-1
      - name: container-2
        resources:
          requests:
            memory: 50Mi
            cpu: 1
  ```

  **Resources requests at pod-level**

  * The scheduler will use the requested resources (CPU and memory) settings specified in the PodSpec to find a suitable node to place the pod. 
  * Kubelet will use the CPU requests from the pod-level resources section, and translate it to cpu.weight in pod-level cgroup.

  **Resources limits at pod-level**

  * All init and non-init containers will be capped by the pod level limit which represents the peak usage of all containers.
  * Kubelet will use CPU and memory limits from PodSpec, and translate them to cpu.max and memory.max cgroup settings at pod-level respectively. 

  **Pros**

  * Simple to understand and implement.
  * Guarantees enough resources for all containers.

  **Cons**

  * Can lead to over-provisioning (resource wastage), especially if the init container peak usage is significantly higher than the steady-state usage 

>   of the other containers. 

Reviewer 2: I think this case can be mitigated by recommending users use container level limits instead. I prefer the logical and programatic simplicity of this option and don't know if this con outweighs those

Reviewer 3: +1 to this. Users requiring more fine-grained control can choose to specify the container-level requests/limits instead. One motivation for having pod-level resource settings is to *simplify* the UX, and we should preserve the simplicity for the users.

>   If this option is picked, it might be worth providing guidance and examples in the feature documentation to help users effectively set pod-level requests and limits. 

Reviewer 3: IMHO, I don't think guidance and examples are needed. We can just call out the difference in the documentation. AFAIK, we don't provide in-depth recommendations on how to set the resource requests/limits, and this is not that different.

  
  Example 1: If your application containers require high resources simultaneously, it might be worth setting pod-level requests/limits such that it is an aggregate of all
  container-level requests/limits.

  Example 2: If your application has init containers with high resource requirements and usage, while your application containers require fairly less resources, it might be
  worth setting pod-level requests/limits as maximum of init and non-init container resources requirements. 

>   Since init containers run sequentially, there is no requirement to share resources with any other container. Therefore setting resource requests and limits as the higher value 

Reviewer: except for the sidecars

Reviewer: expanding on this: sidecar containers will start sequentially and then run while next init containers are running. So resources usage calculation should account for sidecar container + max of init containers running after this sidecar container

        = max (memory: 300Mi cpu: 4, memory: 100 Mi cpu: 2)

        = (memory: 300 Mi, cpu: 4) to find a suitable node.

  * The init containers will rely on its own requests specified in the PodSpec. 
  * After the init containers execution is completed, kubelet will use the pod-level CPU request to set cpu.weight at pod-level cgroup.

  **Resources limits at pod-level**

  * The init containers will rely on their own limits specified in the PodSpec.
  * After the init containers execution is completed, kubelet will use CPU and memory limits at pod-level, and translate them to cpu.max and memory.max cgroup settings
  at pod level respectively. 

  **Pros**

  * Reduces the problem of overprovisioning.

  **Cons**

  * Complicates the definitions of QoS classes as this will require considering requests/limits for both initContainer and pod-level requests/limits to attach a 
  QoS class with a pod. It might be worth setting dynamic QoS classes if this option were to be implemented. The QoS class in the initialization phase is based on
  init containers requests/limits, while the QoS class after the initialization phase takes pod-level requests/limits into account. 

>   * Complicated to implement as it requires additional logic in kubelet to monitor init containers completion after which it will set/change pod level cgroup settings.

Reviewer 7: Another con is that it's confusing as a dev to understand that initContainers follow a different logic and that I have to do the math above to understand the actual cpu/memory request.

**Note: Option 1 seems easier to implement in the alpha phase to gather feedback from the users. If there are enough supporting use cases for Option 2, we can explore 
the implementation in the beta phase of the KEP. Option 2 will also require discussions about dynamic QoS classes as per the execution phase of the pod 
(init vs non-init phase).**

#### Admission Controller

> * LimitRanger

Reviewer: do you want to have a defaulter for Pod resources as well? Or this will be out of scope?

Reviewer 3: Good question. 

Today, in the LimitRanger, you specify the "container" level resource requirements. https://kubernetes.io/docs/tasks/administer-cluster/manage-resources/memory-default-namespace/#create-a-limitrange-and-a-pod
We shouldn't change how that works since it has nothing to do with the pod level resource.

What we can do is to extend the LimitRanger to allow `type: Pod` as well for users who'd like to apply pod level defaults.

Reviewer 6: +1 on what Yuju suggested above.

Author: https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L5673 LimitRanger already supports Pod type today. Reframed this section to add more clarity.

Reviewer: will we implement any logic in limit ranger to account for the pod level requests? If Pod has minimum set, should limit ranger still set minimums on container level? Or it should refrain from updating container-level values?

What if there is a defaulter installed and somebody trying to us this feature. They will be forced to specify some small requests on containers to avoid defaulter overriding those values. Is it OK to force people to set those small values? If so - we need to document it into the rollout section

Reviewer 3: Since limit ranger is implemented as a list, I think it's better to keep different limitRangers independent. 

For alpha, we can suggest users to use a different namespace if they don't want the default limitRanger for containers.

Author: If there's a limit ranger of type container, it will continue to update the container-level resources even if the user uses pod-level resources. If the user doesn't want to use the container-level defaulting, they should create pods in a different namespace.

container-level memory requests to calculate the effective pod-level memory request.When pod-level memory requests are specified, the Eviction Manager should use the pod-level request 
directly instead of aggregating container-level requests. This logic should be modified to check if pod-level requests are set. If so, use the pod-level request; otherwise, fall back 
to aggregating container requests.

#### PodOverhead

Since the PodOverhead feature works at pod-level cgroup, pod-level resource specification and PodOverhead will work together seamlessly. When the kube-scheduler is deciding 
which node should run a new Pod, the scheduler considers that Pod's overhead as well as the sum of container requests for that Pod. Scheduler should add pod-level request,
podoverhead to find a suitable node, when pod level request is specfied.

Likewise the kubelet will need to add PodOverhead to the limits specified in the pod-level spec (instead of sum of container level requests) when setting the upper limit for
the pod-level cgroup (i.e. cpu.max and memory.max).

#### [Scoped for Beta] HugeTLB cgroup

TBD. Do not review for the alpha stage.

#### [Scoped for Beta] Topology Manager

> TBD. Do not review for the alpha stage.

Reviewer: at least high level explanation on how this will be integrated is needed for alpha stage

* (Tentative) Only pod level scope for topology alignment will be supported if pod level requests and limits are specified without container-level requests and limits.

#### [Scoped for Beta] Memory Manager

TBD. Do not review for the alpha stage.

#### [Scoped for Beta] CPU Manager

TBD. Do not review for the alpha stage.

#### [Scoped for Beta] In Place Pod Resize

TBD. Do not review for the alpha stage.

#### [Scoped for Beta] VPA

TBD. Do not review for the alpha stage.

#### [Scoped for Beta] Cluster Autoscaler

TBD. Do not review for the alpha stage.

> t

Reviewer: same comment as above. This is a major change and understanding how things will work together would be needed to approve the API

Author: Added the details.

necessary to implement this enhancement.

#### Unit tests

This feature will touch multiple components. For alpha, unit tests coverage for following packages needs to be added:

* Scheduler logic will be updated to consider pod-level requests. Hence pkg/scheduler will require additional coverage.
* pkg/kubelet/cm will be updated to set pod-level cgroups using CPU requests and limits, and memory limits.
* pkg/apis/core/validation/types_test.go and pkg/apis/core/validation since new fields are added in PodSpec and also new validation rules are required for the new fields.
* pkg/kubeapiserver/admission for changes made in the admission logic for LimitRanger and ResourceQuota admission controllers.

#### e2e tests

Following scenarios need to be covered:

* Cgroup settings when pod-level resources are set.
* Validate scheduling and admission.
* Validate the containers with no limits set are throttled on CPU when CPU usage reaches Pod level CPU limits.
* Validate the containers with no limits set are OOMKilled when memory usage reaches Pod level memory limits.

### Graduation Criteria

#### Phase 1: Alpha (target 1.31)

> * Feature is disabled by default. It is an opt-in feature which can be enabled by setting the new fields in PodSpec.

Reviewer 2: I don't think a feature should be enabled by stting the fields on the object the feature uses. there should be a gate and it should filter the fileds from the API request, like other KEPs

Reviewer 3: +1 on introducing a feature gate. This provides the clear code isolation as well.

Reviewer 5: Yep, should be Feature Gate + pod field

* Support the basic functionality for scheduler to consider pod-level resource requests to find a suitable node.
* Support the basic functionality for kubelet to translate pod-level requests/limits to pod-level cgroup settings.
* Unit test coverage.
* E2E tests.
* Documentation to announce the feature in alpha phase.

#### Phase 2:  Beta (target 1.32)

* User Feedback.
* Feature is disabled by default. It is an opt-in feature which can be enabled by setting the new fields in PodSpec.

> * Support for all features that are scoped out of alpha phase in the KEP design.

Reviewer 3: I'm not sure this needs to be the goal. Would like to hear from others about the convention. Do we need to support all GA features, or all beta features? I think it might be okay if certain features are not in the scope for beta.

Reviewer 5: There could be multiple betas if needed.

Author: Resolving this as we want to support all the features ideally as discussed in the Sig Node meeting today.

Reviewer 3: As mentioned in the meeting, with so many different features in various stages of their individual lifecycles, we may choose to ensure some features are "compatible" with this KEP, instead of "supported" in various stages of this KEP.  As long as it's clear (e.g., from validation) that the combination is currently not supported yet. We can discuss individual features in more details after alpha. 

@dchen1107 @mrunalp to weigh in on this to see if this is compatible with the current convention.

* Unit test coverage.
* E2E tests.

> * Documentation update to announce feature in beta phase.

Reviewer 5: nit: Documentation would add it like any other new feature while a blog could be used to announce/solicit feedback through usage.

> * Benchmark tests for resource usage.

Reviewer 2: these pieces are copied from alpha, do we expect them to change much?

Author: Added the clause to add new e2e tests for features for which the support will be added in Beta.

> * Benchmark tests for resource usage.

Reviewer 3: What kind of benchmark testing do you have in mind?

Author: I was thinking of checking the resource utilization for the workloads similar to two user stories mentioned in the KEP. We could compare the resource usage for with and without pod-level resources.

#### GA (stable)

* TBD

### Upgrade / Downgrade Strategy

The existing workloads will not have any impact since they won't be using the new field `resources` in PodSpec 
at pod-level. If a cluster is upgraded, the cluster administrator will experience pre-upgrade behavior until
`resources` field is specified at pod-level. 

If the workloads are not using the new feature, downgrade won't be affected. However, if it is required to
downgrade a cluster that has workloads using the pod-level limits functionality, following steps need to be
performed:

* Translate pod-level resource specifications to container-level specifications.

* Delete existing workloads and recreate them with container-level specifications.

> 

Reviewer 2: if a user doesn't do these things, then they would loose their limits (as resources on pod spec would be stripped out as unknown by the API server). Is there anything to be done about that?

Reviewer 3: @haircommander I think this is typical for any new feature/field. Do you have any suggestion?

Reviewer 2: yeah I mean if we deemed this enough of an issue a fix would be https://github.com/kubernetes/enhancements/pull/4678#discussion_r1624946950 . I'm not feeling super stubborn about it but it is a concern.

Reviewer 3: I think we can clearly document that toggling the feature may leave some pods in the undesired state, and to address the issues, users can delete existing pods. WDYT?

* Drain the nodes one by one to prevent the new pods from being scheduled on them.

* Downgrade the kubelet on the drained nodes to a version that doesn't support pod-level resource specification
fields.

* Downgrade the control plane components.

* Uncordon the nodes, making it available for scheduling again.

### Version Skew Strategy

Pod-level resource specification is an opt-in feature that gets enabled by supplying the `resources` stanza 
in the `spec` at pod level. For this feature to work correctly, it must be enabled in all parts of the 
cluster (scheduler, API server, kubelet). However, if a node is running an older version of kubelet that 
doesn't support this feature, it will ignore the pod-level resoucr specifications and continue to use the 

> aggregated container-level values (backward compatible). Applications must be prepared for this version skew 

Reviewer 2: maybe during the transition time we should require users specify container limits in the apiserver until all supported kubelets support this field, to prevent the risk of a user not specifying any limits and messing up scheduling/runtime of the application

Reviewer 3: For some users, they'd like to specify pod level resources so that they don't have to deal with individual container resources, so this'd defeat the purpose for them. 

This also seems to be a common issue we'd face for similar features. Maybe we can follow the convention?

Reviewer 2: yeah totally, and I think that's a good end-goal, but requiring them in the short term to specify container limits despite the pod limits being in place would ensure their pods are in the right QOS class and are limited/schedule correctly

    This feature can be enabled by adding `resources` in the `spec` at pod-level.
    Following components depend on it:
    - kubelet
    - kube-apiserver
    - kube-scheduler
  - Will enabling / disabling the feature require downtime of the control
    plane? No. `resources` must be set/unset in `spec` at pod-level to enable/disable this feature.
    The workloads need to restart to disable the feature.
  - Will enabling / disabling the feature require downtime or reprovisioning
    of a node? No. Only the workloads need to restart.

###### Does enabling the feature change any default behavior?

No.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes. Pods created with pod level resource spec need to be deleted and recreated with only container level resources.

###### What happens if we reenable the feature if it was previously rolled back?

The pods will go in Pending state if the feature is rolled back due to "unknown" `resources`
field at pod-level. It will become schedulable again once the feature is reenabled making pods
schedulable again.

###### Are there any tests for feature enablement/disablement?

> Yes, the tests will be added alongwith alpha implementation.

Reviewer 2: along with*

<!--
Look at the [existing SLIs/SLOs].

Think about adding additional work or introducing new steps in between
(e.g. need to do X to start a container), etc. Please describe the details.

[existing SLIs/SLOs]: https://git.k8s.io/community/sig-scalability/slos/slos.md#kubernetes-slisslos
-->

The computational overhead introduced in API server, scheduler and kubelet due to additional validation for pod level resource specifications should be negligible. Thorough testing and monitoring will ensure the SLIs/SLOs are not impacted.

###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?

<!--
Things to keep in mind include: additional in-memory state, additional
non-trivial computations, excessive access to disks (including increased log
volume), significant amount of data sent and/or received over network, etc.
This through this both in small and large cases, again with respect to the
[supported limits].

[supported limits]: https://git.k8s.io/community//sig-scalability/configs-and-limits/thresholds.md
-->

> No. Pod-level resource information could lead to marginal increase in storage requirements for the new fieleds, although

Reviewer 2: fields *

its impact is likely to be small compared to other Kubernetes object data.

###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?

<!--
Focus not just on happy cases, but primarily on more pathological cases
(e.g. probes taking a minute instead of milliseconds, failed pods consuming resources, etc.).
If any of the resources can be exhausted, how this is mitigated with the existing limits
(e.g. pods per node) or new limits added by this KEP?

Are there any tests that were run/should be run to understand performance characteristics better
and validate the declared limits?
-->
Yes. With pod-level resource limits, it's possible to pack more pods onto a node. This increased
pod density can also lead to higher consumption of resources like PIDs and sockets, as each
pod and its containers create processes and potentially establish network resources.

> 

Reviewer 2: I would also say that the maxPods field should already account for and limit the problems of this

Reviewer: was it about mode pods or more containers? I wonder if we will see a growing number of Pods with many containers. Since you don't have to specify requests and limits on individual containers, people may start packing more low traffic containers into Pods. But I think it is more suitable for the risks section

      running user workloads?
    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->

###### What steps should be taken if SLOs are not being met to determine the problem?

It is suggested that if workloads with pod level resource specifications enabled cause performance 
or stability degradations, those workloads are recrated with only container level resource specs.

## Implementation History

- ** 2020-03-17:** Initial discussion in [SIG Node meeting] (https://www.youtube.com/watch?v=3cU56ZiUZ8w&list=PL69nYSiGNLP1wJPj5DYWXjiArF-MJ5fNG&index=101)
- ** 2020-03-30:** Initial [KEP draft](https://github.com/kubernetes/enhancements/pull/1592) and discussion.
- ** 2021-07-26:** Issue [link](https://github.com/kubernetes/enhancements/issues/2837)
- ** 2024-05-31:** Revised KEP for alpha (#4678)[https://github.com/kubernetes/enhancements/pull/4678]

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

> ## Alternatives

Reviewer: there is no really a good alternative for this, are there? Please tell so here

Reviewer 3: For simplifying the user experience, pod resizing could be one alternative. But you need to integrate that with autoscaling, and there's still additional delay that's needed to adjust the resources. There's also the risk of the resizing request being denied.

### Risks and Mitigations

Since this is an **opt-in** feature, there should be no risk in merging the changes. Existing workloads that do not use this feature won’t be affected.
Besides, the feature will be introduced in a phased approach, starting with an alpha release for testing and gathering feedback, followed by beta and GA releases
as the feature matures and potential issues are addressed.

## Design Details

### Components/Features requiring changes

#### PodSpec and PodStatus changes

New field in `PodSpec` and `PodStatus`

```
type PodSpec struct {
...
  // Compute resource requirements.
  // +optional
  Resources ResourceRequirements
}

type PodStatus struct {
...
  // Compute resource requirements.
  // +optional
  Resources ResourceRequirements
}
```

#### PodSpec Validation

* **Scenario 1: Only container-level resources are specified (which is already supported without this KEP)**

  ```yaml
  apiVersion: v1
  kind: Pod
  ...
  spec:
    containers:
    - name: container-1
      resources:
        requests:
          memory: 100Mi
          cpu: 1
    - name: container-2
  ...
  ```

  No changes required to support this.

> * **Scenario 2: Only pod-level resources are specified**

Reviewer 3: This is discussed in the LimitRange section below.

> * **Scenario 2: Only pod-level resources are specified**

Author: Container-level resource defaulting is preserved in the updated proposal.

  ```yaml
  apiVersion: v1
  kind: Pod
  spec:
    resources:
      requests:
        memory: 100Mi
        cpu: 1
  ...
    containers:
    - name: container-1
    ...
    - name: container-2
  ...
  ```

  Following validation logic needs to be added for `resources` specified in the PodSpec:

  1. Only CPU and memory resources are supported in pod level resource specifications, hence invalidate pod-level resource specification for other resources 
  like gpu, storage, etc.

* **Scenario 3: Both pod-level and container level resources are specified.**

  ```yaml
  apiVersion: v1
  kind: Pod
  ...
  spec:
  resources:
    requests:
      memory: 100Mi
      cpu: 1
    containers:
    - name: container-1
        resources:
          requests:
            memory: 100Mi
            cpu: 1
    ...
    - name: container-2
  ...
  ```

  Following validation logic needs to be added:

  1. requests

>       * `pod.spec.resources.requests[memory/cpu] >= sum(pod.spec.containers[*].resources.requests[memory/cpu])` is valid.

Reviewer 4: Does this mean "calculate the implicit pod request, in the standard way, and ensure that pod requests is at least as much" ?

>       * `pod.spec.resources.requests[memory/cpu] >= sum(pod.spec.containers[*].resources.requests[memory/cpu])` is valid.

Reviewer 8: Yeah, that was my interpretation.

>       * `pod.spec.resources.requests[memory/cpu] >= sum(pod.spec.containers[*].resources.requests[memory/cpu])` is valid.

Reviewer 3: I think this can section benefit from defining some terms for clarity, e.g., 
 - "Minimum effective pod request" for the implicit pod request from all the init/sidecar/app containers
 - `pod.spec.resources.requests[memory/cpu] >= minimum_effective_pod_request`

          all the containers within a pod to function correctly. If a pod’s request is lower than the sum of its container requests, it creates a logical 
          contradiction: the pod as a whole requires less resources than the sum of its parts.

          Besides, the scheduler will use pod requests to determine suitable nodes for placement. If a pod’s request underestimates its actual resource needs
          (represented by the sum of container requests), it could be scheduled on a node that cannot provide enough resources, leading to failures and 
          performance degradation.

  2. limits
      * `pod.spec.resources.limits[memory/cpu] >= sum(pod.spec.containers[*].resources.limits[memory/cpu])` is valid.

          Pod-level resource limits are allowed to be greater than the sum of container level resource limits to account for ephemeral containers that could be added 
          later. Setting a resource limit on the pod in its cgroup settings will not reserve any resource and hence not affect the other workloads running on the node.

>       * `pod.spec.resources.limits[memory/cpu] < sum(pod.spec.containers[*].resources.limits[memory/cpu])` is valid.

Reviewer 4: That should already be true in all cases

>       * `pod.spec.resources.limits[memory/cpu] < sum(pod.spec.containers[*].resources.limits[memory/cpu])` is valid.

Reviewer 8: I don't think so, since pod limits are allowed to be smaller than the sum of container limits. If you set a pod limit, and omit pod requests, I think you could end up with this:

pod limit: 10
pod requests: nil
container 1 limit: 10
container 1 request: 10
container 2 limit: 10
container 2 request: 10

If we required pod requests to be set whenever pod limits are set, then I think it holds.

>       * `pod.spec.resources.limits[memory/cpu] < sum(pod.spec.containers[*].resources.limits[memory/cpu])` is valid.

Reviewer 3: Something like `pod.spec.resources.limits[memory/cpu] >= minimum_effective_pod_requests[memory/cpu]`

>       * `pod.spec.resources.limits[memory/cpu] < sum(pod.spec.containers[*].resources.limits[memory/cpu])` is valid.

Reviewer 3: > When request is not specified, it defaults to the limit, right?

yes, that's true for containers and we should apply the same logic to pods. @ndixita

  * After the init containers execution is completed, kubelet will use the pod-level CPU request to set cpu.weight at pod-level cgroup.

  **Resources limits at pod-level**

  * The init containers will rely on their own limits specified in the PodSpec.
  * After the init containers execution is completed, kubelet will use CPU and memory limits at pod-level, and translate them to cpu.max and memory.max cgroup settings
  at pod level respectively. 

  **Pros**

  * Reduces the problem of overprovisioning.

  **Cons**

  * Complicates the definitions of QoS classes as this will require considering requests/limits for both initContainer and pod-level requests/limits to attach a 
  QoS class with a pod. It might be worth setting dynamic QoS classes if this option were to be implemented. The QoS class in the initialization phase is based on
  init containers requests/limits, while the QoS class after the initialization phase takes pod-level requests/limits into account. 
  * Complicated to implement as it requires additional logic in kubelet to monitor init containers completion after which it will set/change pod level cgroup settings.

> **Note: Option 1 is a *preferred* design option as it is easier to implement, and also easier to understand from UX perspective. 

Reviewer 4: Agree - I think option 2 is too hard to reason about - principle of least surprise.

If there are enough supporting use cases for Option 2, we can explore the implementation in the beta phase of the KEP. Option 2
will also require discussions about dynamic QoS classes as per the execution phase of the pod (init vs non-init phase).**

#### Admission Controller

* LimitRanger

>   This admission controller applies default resource requests to the Pods. In the existing implementation (without this KEP), it sums up container-level requests to calculate

Reviewer 3: "default resource requests to the Pods." --> This is applied to individual containers in a pod, right?

Author: PTAL, I reworded the entire admission controller section with more refined details.

that do utilize this feature have poorly configured limits, they could become noisy neighbours and
negatively impact other workloads on the same node. It's important to nose that this issue of noisy
neighbours due to incorrect limits it not exclusive to pod-level resources; it can also occur
at with container-level limits.

If a cluster is upgraded, the cluster administrator will experience pre-upgrade behavior until 
`resources` field is specified at pod-level. 

If the workloads are not using the new feature, downgrade won't be affected. However, if it is required to
downgrade a cluster that has workloads using the pod-level limits functionality, following steps need to be
performed:

* Translate pod-level resource specifications to container-level specifications.

* Delete existing workloads and recreate them with container-level specifications.

* Drain the nodes one by one to prevent the new pods from being scheduled on them.

* Downgrade the kubelet on the drained nodes to a version that doesn't support pod-level resource specification
fields.

* Downgrade the control plane components.

* Uncordon the nodes, making it available for scheduling again.

> ### Version Skew Strategy

Reviewer 8: I would add: if the feature gate is disabled, but the pod-level resources field is set, the Kubelet should reject the pod.

It won't matter much for the alpha case, but it will help with the alpha->beta transition.

Alternatively, we could say that the kubelet always honors the field (basically ignore the feature gate), and only control the feature at the cluster level.

-->

###### How can this feature be enabled / disabled in a live cluster?

<!--
Pick one of these and delete the rest.

Documentation is available on [feature gate lifecycle] and expectations, as
well as the [existing list] of feature gates.

[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
-->

- [X] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: PodLevelResources
  - Components depending on the feature gate: kubelet, kube-apiserver, kube-scheduler
  - Will enabling / disabling the feature require downtime of the control
    plane? No. `resources` must be set/unset in `spec` at pod-level to enable/disable this feature.
    The workloads need to restart to disable the feature.
  - Will enabling / disabling the feature require downtime or reprovisioning
    of a node? No. Only the workloads need to restart.

###### Does enabling the feature change any default behavior?

Yes, if the feature is enabled and pod level resources are specified the scheduler and kubelet will consider

> pod level values over container level resource specs. 

Reviewer 8: I think this doesn't count as default behavior, since the user needs to set a new field for it to work. If the user doesn't set pod level resources, there is no change to behavior, right?

mid-rollout?

Be sure to consider highly-available clusters, where, for example,
feature flags will be enabled on some API servers and not others during the
rollout. Similarly, consider large clusters and how enablement/disablement
will rollout across nodes.
-->

Since the feature is opt-in feature and requires setting new fields in the PodSpec,
the rollouts won't likely fail because of this feature. For the new workloads
that want to use this feature, there could be unexpected interactions with
the existing features. This is why this feature will be rolled out in phases
to make sure all the cases and interaction with existing features are covered
before making it available in GA.

Rollbacks should be seamless if done after disabling the feature, and recreating 
the running workloads that use the feature.

###### What specific metrics should inform a rollback?

<!--
What signals should users be paying attention to when the feature is young
that might indicate a serious problem?
-->

The following metrics could indicate problems with API server validation failures, latency or requests dropped by API server, scheduling latencies, etc:

> * `apiserver_request_total`

Reviewer 8: duplicate

* `apiserver_request_duration_seconds`

> * `apiserver_request_total`

Reviewer 8: These don't seem that relevant?

1. Community is trying to move cgroupv1 to maintenance mode in future Kubernetes versions (likely 1.31). Hence this feature will be supported only on cgroupv2.
2. This feature is an **opt-in** feature, and will have no effect on existing deployments. Only deployments that explicitly require this functionality should
turn it on by specifying the relevant resource section at pod-level in the Pod specification.

### Risks and Mitigations

Since this is an **opt-in** feature, there should be no risk in merging the changes. Existing workloads that do not use this feature won’t be affected.
Besides, the feature will be introduced in a phased approach, starting with an alpha release for testing and gathering feedback, followed by beta and GA releases
as the feature matures and potential issues are addressed.

## Design Details

### Components/Features requiring changes

#### PodSpec and PodStatus changes

New field in `PodSpec` and `PodStatus`

```
type PodSpec struct {
...
  // Compute resource requirements.
  // +optional
  Resources ResourceRequirements
}

type PodStatus struct {
...
  // Compute resource requirements.
  // +optional
  Resources ResourceRequirements

> }

Reviewer 8: I don't see any discussion of the pod status. Why is there a new field for pod status? I assume it is for pod resize? But that's out of scope for alpha, and missing the allocated resources.

Can this be omitted for now?

Author: Correct. Omitting it for now.

rollout. Similarly, consider large clusters and how enablement/disablement
will rollout across nodes.
-->

Since the feature is opt-in feature and requires setting new fields in the PodSpec,
the rollouts won't likely fail because of this feature. For the new workloads
that want to use this feature, there could be unexpected interactions with
the existing features. This is why this feature will be rolled out in phases
to make sure all the cases and interaction with existing features are covered
before making it available in GA.

Rollbacks should be seamless if done after disabling the feature, and recreating 
the running workloads that use the feature.

###### What specific metrics should inform a rollback?

<!--
What signals should users be paying attention to when the feature is young
that might indicate a serious problem?
-->

The following metrics could indicate problems with API server validation failures, latency or requests dropped by API server, scheduling latencies, etc:

* `apiserver_request_total`
* `apiserver_request_duration_seconds`
* `apiserver_request_total`
* `scheduler_scheduling_duration_seconds`

> * `scheduler_pod_unschedulable_count`

Reviewer 8: Are there any relevant node-level metrics?

## Motivation

Kubernetes workloads often consist of multiple containers collaborating within a pod. While this separation enables granular monitoring and management of each
container, it can be challenging to accurately estimate and allocate resources for individual containers, especially for workloads with unpredictable or 
bursty resource demands. This often leads to over-allocation of resources to ensure that no container experiences resource starvation, as kubernetes currently
lacks a mechanism for sharing resources between containers easily. 

Introducing pod-level resource requests and limits offers a more simplified resource management for multi-container pods as it is easier to gauge the collective
resource usage of all containers in a pod rather than predicting each container’s individual needs. This proposal aims to support pod-level resource requests 
and limits that would allow Kubernetes to manage the total resource consumption of the pod, freeing users from the burden of meticulously allocating resources
for each container. This simplified approach is particularly valuable for tightly coupled applications, and it will enable Kubernetes to leverage the underlying

> cgroup resource management mechanisms more effectively for better resource utilization for bursty workloads. 

Reviewer 4: cgroups is how linux works - If this covers windows, too, then it should maybe mention that up front.

Author: Added a section for this.

Additional tools are provided in additional side-car deployments - for example an LSP( https://langserver.org/) service, a terminal, and more. 
Using this new feature, the containers providing the terminal, the LSP services, and the set of tools being utilized can share the resource limit
defined for the pod. Consider the following pod definition:

```yaml
apiVersion: v1
kind: Pod
metadata:
  labels:
    run: ide
  name: myide
spec:
  resources:
    limits:
      memory: "1024Mi"
      cpu: "4"
  containers:
  - image: shell
    name: debian:buster
  - image: tool1
    name: first-tool:latest
  - image: tool2
    name: second-tool:latest
  - name: ide
    image: theia:latest
    resources:
      requests:
        memory: 128M
        cpu: "0.5"
      limits:
        memory: 256M
        cpu: "1"
```

Using the pod-level resource definition enables the `tool1` and `tool2` containers to be constrained by the total limit for the pod. 

Without this feature, the developer would need to decide a-priori how much resources to allocate to the tools - and this is not easy to do for this workload.

#### Story 2

> Consider a scenario where a pod has some containers with occasional resource spikes while other containers often have idle resources. With current container 

Reviewer 4: This story is kind of weak, and a bit misleading.  I think it needs to emphasize that it saves resources when you know that resource bursts are not coordinated.  Rather than planning for the sum of a percentile of the max of each container, you can plan for a percentile of the aggregate, which should be lower than the sum of individuals' max.

minor point

Reviewer: +1 on adding this clarification. Also to make case stronger - example of service mesh sidecar can be added. If sidecar that provides the service mesh is CPU throttled, it makes the whole Pod non-functional even if it has CPU cycles. So it only logical to make service mesh share resources with the main container in many cases

  kind: Pod
  spec:
    resources:
      requests:
        memory: 100Mi
        cpu: 1
  ...
    containers:
    - name: container-1
    ...
    - name: container-2
  ...
  ```

  Following validation logic needs to be added for `resources` specified in the PodSpec:

  1. Only CPU and memory resources are supported in pod level resource specifications, hence invalidate pod-level resource specification for other resources 
  like gpu, storage, etc.

* **Scenario 3: Both pod-level and container level resources are specified.**

  ```yaml
  apiVersion: v1
  kind: Pod
  ...
  spec:
  resources:
    requests:
      memory: 100Mi
      cpu: 1
    containers:
    - name: container-1
        resources:
          requests:
            memory: 100Mi
            cpu: 1
    ...
    - name: container-2
  ...
  ```

  Following validation logic needs to be added:

  1. requests
      * `pod.spec.resources.requests[memory/cpu] >= sum(pod.spec.containers[*].resources.requests[memory/cpu])` is valid.

        Pod-level resource requests are allowed to be greater than the sum of container level resource requests to account for ephemeral containers that could 
        be added later.

>       * `pod.spec.resources.requests[memory/cpu] < sum(pod.spec.containers[*].resources.requests[memory/cpu])` is invalid.

Reviewer 4: Isn't this just the exact opposite of the above?  Or am I missing some nuance?  I don't think we need to spec both.  It reads as "X is true and !X is false" - that's tautological.

Reviewer 3: +1. doesn't need to re-iterate this again

          1. Container 1: Moderate CPU and memory usage, but occasional spikes.
          2. Container 2: Low and consistent resource usage.

          ```yaml
          apiVersion: v1
          kind: Pod
          ...
          spec:
            resources:
              requests:
                memory: 100Mi
                cpu: 1
              limits:
                memory: 400Mi
                cpu: 2.5
            containers:
            - name: container-1
                resources:
                  requests:
                    memory: 60Mi
                    cpu: 0.5
                  limits:
                    memory: 300Mi
                    cpu: 2  
            - name: container-2
                resources:
                  requests:
                    memory: 40Mi
                    cpu: 0.5
                  limits:
                    memory: 200Mi
                    cpu: 1 
          ```

          The pod-level limit will serve as an aggregate ceiling. This will cover the scenario where the sum of peak usage for individual containers in a pod is high, but these
          peaks don’t necessarily occur simultaneously.

>           If we simply set the pod-level limit as the sum of the individual container limits, we might overallocate resources. The pod might never actually use all those resources

Reviewer 4: FWIW, I think this is the most interesting aspect of the whole KEP. 

If I make a pod which requests.memory = 8Gi and limits.memory = 8Gi, and I have 2 containers which each requests.memory=4Gi and limits.memory = 5Gi, I think think the meaning is clear and unambiguous.  I want guaranteed QOS.  I will not over-commit the machine.  I am asserting that my own containers might burst by 25%, but I think they will rarely burst at the same time.  And if they do, I should expect reclaim and OOM inside my pod.

Reviewer: Another flavor of the same: https://github.com/kubernetes/enhancements/pull/4678/files#r1645182074

          at once, leading to wasted capacity on the node.

      * `pod.spec.resources.limits[memory/cpu] < pod.spec.containers[i].resources.limits[memory/cpu]` is invalid.

        A single container’s limit should not be allowed to exceed the pod level limit.

>   3. Validation logic comparing requests and limits

Reviewer 4: I'd like to see an explanation or table showing what happens if a user does not specify each of requests/limits x cpu/mem x pod/container.

Today we have some runtime-default minimums, and we have rules like the default limit is infinity, but the default request is limit.  How does that change?

Reviewer 3: +1 for having a table. 

@thockin, for defaults, are you talking about LimitRanger or something else? There's a section below to talk about LimitRanger, but it may need to be expanded a little bit to cover all the cases.

Author: Added the table. PTAL and review

      3. Sum of container level requests greater than Pod level limits is invalid.
        
          This is already ensured by other validation rules in 1. and 2. as following:
        
          As per 1. Container level Requests <= Pod level requests is valid<br>
          And as per 3.1 Pod-level requests <= Pod level limits is valid<br>
          Hence, by transitivity these rules will ensure Container level requests <= Pod-level limits.

      4. [Already exists] Sum of container level requests less than container level limits is invalid.

#### Scheduler Changes

The scheduler logic needs to be updated to support pod-level requests and limits.

* **Scenario 1: Only container-level resources are specified**

  No changes, same as existing.

* **Scenario 2: Only pod-level resources are specified**

  Consider pod-level requests to find the suitable node for pod placement.

* **Scenario 3: Both container-level and pod-level resources are specified**

  Scheduler will prioritize pod-level requests as the sole indicator of the pod’s resources needs to find a node that will fulfill the overall 
  resource requirements of the pod. It will ignore container level requests.

> #### QoS Changes

Reviewer 4: Similar to above, this whole section feels like it could be reduced to something like "QoS is determined by the derived pod resources, which considers the pod-level resource stanza and all the containers' resources stanzas."

Reviewer 5: +1 agree

    2. Implicit Pod-level Resources (via containers)
       * Pod level `resources` setting is not specified.
       * *Every* container within the pod has requests and limits specified for each resource.
       * All container resource requests and limits are equal.

  **Note: If the pod spec includes `resources` stanza at pod level, but either requessts or limits are missing for a resource, the pod will not be considered
  Guaranteed QoS, even if container-level requests and limits are present and are equal.**

* **Burstable QoS Criteria**

    A pod can qualify for Burstable QoS in one of the following scenarios:

    1. Explicit Pod-level Resources

       * It doesn't meet the criteria of Guaranteed QoS. 
       * At least one of pod-level requests or limits is specified for CPU or memory.

    2. Implicit Pod-level Resources (via containers)

       * It doesn't meet the criteria of Guaranteed QoS. 
       * Pod level `resources` setting is not specfied.
       * At least one Container in the Pod has a memory or CPU request or limit.

* **BestEffort QoS Criteria**

    None of pod-level or container-level resources requests or limits are specified.

> #### Init Containers & Sidecar Containers

Reviewer 4: You should put this section earlier, and then define sceduler and QoS behavior in terms of it, I think?

      to specify default requests and limits for pods in a namespace. This requires changing the 
      (validation logic)[https://github.com/kubernetes/kubernetes/blob/9e2075b3c87061d25759b0ad112266f03601afd8/pkg/apis/core/validation/validation.go#L6403]
      to allow `defaultRequest` and `default` in Pod LimitRange object.

      * Conditional logic
        
         * Pod level resource spec is set and `Pod` LimitRanger is applied.
           If pod level resources are specified, use these values directly against the Pod type LimitRanger. It doesn't matter if container level
           resource spec is set or not.
           
         * Pod level resource spec is set and `Pod` LimitRanger is applied.
           If pod level resources are not specified, there could be cases:
             1. Existing workloads that have `Pod` LimitRanger set to validate the constraints at container level aggregates (and not at pod level).
             2. New workloads that want to use Pod Level resource limits feature and have set `Pod` LimitRanger to validate constraints at Pod Level.
           

>            If the behavior of `Pod` LimitRanger is modified, it could break existing workloads and also remove the ability for usecase where users 

Reviewer 4: I am unclear what breaks here?  It seems like this is the OBVIOUS extension, no?  

We already have an "effective" pod resources, as detailed above.  What does `LimitRange` do for `type: Pod` today -- I would assume (bad, I know) that it checks against that effective value, am I wrong?

Author: LimitRange of type "Pod" already exists today.

LimitRange of type Pod sums up the resources specified for containers as effective values for pod. It is easy to change this to use pod level values with this KEP. But the problem is, today it doesn't support specifying defaultRequest, and default (which is default limit) at pod level since there's no way (without this KEP) to specify these values at Pod level.

Now with the support for defaultRequest and limit in this KEP, these values will be applied when pod level resources are not specified. How to differentiate between when workloads do not specify pod level values so that they can use default values from limit ranger vs when pod level values are not specified for workloads that want to use container level limits only (like in existing workloads)? I might have overcomplicated this, but I wanted to discuss this scenario.

On the contrary, we could say that since defaultRequest and limit can be specified only after pod level resources feature is supported (since today there is no way to specify pod level resources and hence defaultRequest and limits are not supported without this feature), it essentially means that user wants to apply defaultRequest and limit at pod level, if specified in the LimitRange object of type Pod, for all workloads in that namespace including existing workloads. Hence we could leave it to the users to make sure they set defaultRequest and limit only in those namespaces in which they want to support these values as pod level.

           want to have limit range constraints for container level aggregates.         

  * **Option 2: Add a new LimitRanger type PodLevelResources**
    * Introduce a new type `PodLevelResources` in the LimitRange specifically for defining pod-level resource constraints (min, max, defaultRequest, default, maxLimitRequestRatio).
  
    This new type would be independent of the existing Pod and Container LimitRange types with clearer semantics. It maintains backward compatibility as it doesn't modify
    existing Pod type. User education will be needed to clarify the difference between Pod and PodLevelResources types. Plus this is confusing with Pod type LimitRange.

>   * **[Preferred] Option 3: Add a new optional field `Scope` in LimitRangeSpec**

Reviewer 4: I strongly prefer using the field that already describers itself as "Pod" if we can. I'd like to understand why that can't work.

Reviewer 3: +1. 

It should work. And since we disallow setting the default values for pods today, opening this up should not break anyone. 

@ndixita to update the section

    * Introduce a new optional field in LimitRangeSpec with values: Pod or Container(default).

      ```
        // LimitRangeSpec defines a min/max usage limit for resources that match on kind
        type LimitRangeSpec struct {
	      
          // Limits is the list of LimitRangeItem objects that are enforced
	        Limits []LimitRangeItem
          // Allowed values: "Container" (default) or "Pod"
          Scope LimitRangeScope `json:"scope,omitempty"
        }
      ```
      
      The validation logic will use Pod level resource spec directly if scope is set to Pod, otherwise it will apply existing logic for aggregated container resources.

    * Support `default` and `defaultRequest` constraints if scope is "Pod".  

    This option maintains backward compatibility and offers flexibility, allowing both pod-level and container-level constraints in the same LimitRange object. The implementation
    is much simpler than Option 2. Although some user education through docs is still required with this option.
  

> * ResourceQuota

Reviewer 4: Similar to above, this whole section feels like it could be reduced to something like "Quota will consider the effective resource requests and limits for the pod, including the new pod-level resources." ?

means unlimited, an unset hugetlb limit is the same as setting it to 0.

With the proposed changes, hugepages-2Mi and hugepages-1Gi will be added to the pod-level resources section, alongside CPU and memory. The hugetlb cgroup for the
pod will then directly reflect the pod-level hugepage limits, rather than using an aggregated value from container limits. When scheduling, the scheduler will 
consider hugepage requests at the pod level to find nodes with enough available resources.

#### [Scoped for Beta] Topology Manager

Note: This section includes only high level overview; Design details will be added in Beta stage.

* (Tentative) Only pod level scope for topology alignment will be supported if pod level requests and limits are specified without container-level requests and limits.
* The pod level scope for topology aligntment will consider pod level requests and limits instead of container level aggregates.
* The hint providers will consider pod level requests and limits instead of container level aggregates.

#### [Scoped for Beta] Memory Manager

TBD. Do not review for the alpha stage.

#### [Scoped for Beta] CPU Manager

TBD. Do not review for the alpha stage.

> #### [Scoped for Beta] In Place Pod Resize

Reviewer 4: To do in-place resize, we almost certainly need status fields added, and new fields always start as alpha.  So our choices are: do it in this alpha (or at least add the fields), do in in 1.32 but stay alpha, or wait for this to go beta and then add another alpha feature-gate to cover in-place updates of pod-level resources.

@tallclair

Reviewer 3: @tallclair not sure if we ended up discussing this. If we haven't, we should finalize this ...

<!--
Pick one of these and delete the rest.

Documentation is available on [feature gate lifecycle] and expectations, as
well as the [existing list] of feature gates.

[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
-->

- [X] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: PodLevelResources
  - Components depending on the feature gate: kubelet, kube-apiserver, kube-scheduler
  - Will enabling / disabling the feature require downtime of the control
    plane? No. `resources` must be set/unset in `spec` at pod-level to enable/disable this feature.
    The workloads need to restart to disable the feature.
  - Will enabling / disabling the feature require downtime or reprovisioning
    of a node? No. Only the workloads need to restart.

###### Does enabling the feature change any default behavior?

No. This feature is guarded by a feature gate, and requires setting Pod level `resource` stanza explicitly.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

> Yes. Pods created with pod level resource spec need to be deleted and recreated with only container level resources.

Reviewer 9: ```suggestion
Yes. Once the feature gates are disabled, Pods created with pod level resource spec need to be deleted and recreated with only container level resources.
```

Is this correct?

###### What happens if we reenable the feature if it was previously rolled back?

The pods will go in Pending state if the feature is rolled back due to "unknown" `resources`
field at pod-level. It will become schedulable again once the feature is reenabled making pods

> schedulable again.

Reviewer 9: The question here is specifically what happens when we _reenable_ the feature.  Won't it start working again once the workload is restarted?

Re: Pending state. Is this what we want?  We only need to support rollback from beta to alpha. Can we implement alpha such that a a rollback results in pods running without the resources and not getting stuck in Pending state?  Or am I mis-understanding what's being said here?

Author: Correct, if the rollback is done properly as per the plan stated in the rollback section, the pods will run as usual on reenabling the feature.

###### Are there any tests for feature enablement/disablement?

> Yes, the tests will be added along with alpha implementation.

Reviewer 9: Can we enumerate the key tests?  I'm expecting something like:

- When flag is disabled on the kubelet, newly started pods with `resources` ignore the field value, and existing pods with `resources` ignore the field value once restarted
- When the flag is disabled on the scheduler, ?
- When the flag is disabled on the kube-apiserver, existing `resources` field values remain, but attempts to write values to the field are rejected, except removals of the field value, which are allowed..  
- ??

Without this feature, the developer would need to decide a-priori how much resources to allocate to the tools - and this is not easy to do for this workload.

#### Story 2

Consider a scenario where a pod has some containers with occasional resource spikes while other containers often have idle resources. With current container 
level resource settings, the peak usage of each container needs to be accounted for when setting limits. This leads to wasted resources when the spikes are 
not occurring. Pod-level resource limits will allow setting a limit for the entire pod, enabling containers to share resources more efficiently without the
need to over-provision for each container’s peak usage. 

### Notes/Constraints/Caveats

1. Community is trying to move cgroupv1 to maintenance mode in future Kubernetes versions (likely 1.31). Hence this feature will be supported only on cgroupv2.
2. This feature is an **opt-in** feature, and will have no effect on existing deployments. Only deployments that explicitly require this functionality should
turn it on by specifying the relevant resource section at pod-level in the Pod specification.

### Risks and Mitigations

> Since this is an **opt-in** feature, there should be no risk in merging the changes. Existing workloads that do not use this feature won’t be affected.

Reviewer 10: From scheduling and queueing perspective, this will impact all kinds of schedulers and queueing systems(kueue etc.), so this is a potential risk I think or something we should highlight.

The default scheduler depends on the `PodRequests` right now, https://github.com/kubernetes/kubernetes/blob/a81ea5d0af41a7e519a06780aced3c517f61547c/pkg/api/v1/resource/helpers.go#L50, maybe we should expose the function for widely usage to avoid the redundant work.

cc @alculquicondor

Reviewer 3: Good callout. @alculquicondor could you help review the KEP for sig-scheduling? 

For alpha, it should be okay even if other schedulers/queuing systems don't support the feature. But for beta/GA, it'd be good if we can work out a plan for others to easily support this.

Reviewer 10: The most related part is the nodeResource plugin, for null-request Pods, we will apply with a default value for each container to avoid scheduling too many null-request Pods because they're consuming resources actually. This will be conflict with the Pod level requests.

We should change the logic here.

Reviewer 11: Previous discussions around homogenizing the calculation function are in the PR https://github.com/kubernetes/kubernetes/pull/115367

Reviewer 11: ah, there was actually this open PR https://github.com/kubernetes/kubernetes/pull/124609

Reviewer 11: More defaulting logic that we need to take into account https://github.com/kubernetes/kubernetes/blob/f248c24456254d6e0668a424a31258e6d9e9c185/pkg/scheduler/framework/types.go#L927

Besides, the feature will be introduced in a phased approach, starting with an alpha release for testing and gathering feedback, followed by beta and GA releases
as the feature matures and potential issues are addressed.

> ## Design Details

Reviewer: @kad asked to explain how pod requests an limits will be passed to CRI on pod creation to work on alternative runtime handlers. I believe there will be a difference in behavior if VMs are used as implementation. Have you had a chance to follow up on this? Will any CRI changes be needed?

Author: This comment was addressed. There's no change in what values are passed to CRI with Pod level resources. This KEP changes how pod level requests and limits are calculated.

### Components/Features changes

#### PodSpec changes

New field in `PodSpec`

```
type PodSpec struct {
...
  // Compute resource requirements.
  // +optional
  Resources ResourceRequirements
}
```

> #### PodSpec Validation

Reviewer: third column - is request calculated as a effective limit? Or this is a typo?

> #### PodSpec Validation

Reviewer: third column is concerning a bit.

I would expect that customer will be able to say:

I need that much memory for the entire Pod. And I want to limit individual containers to some high value as a precausion. But than - container requests will be equal to container limits. And sum of requests for containers will be higher than Pod request. Which will fail the pod admission as I understand

Reviewer 3: Is it the 3rd row that you were talking about? 

I'm not sure I follow the use case. Individual container requests should represent the expected baseline of the resource consumption, and their sum should not exceed the pod requests.

If user wants to have container request == limit to have the guaranteed class behavior, they'd most likely want to have a corresponding high pod resource request and limit.

Could you elaborate more?

Author: Sum of requests for containers are now not allowed to be greater than pod-level requests.

> <table>

Reviewer: markdown table may be easier to review

   <td>Specified
<p>
Container Limit
   </td>
  </tr>
  <tr>
   <td>Unspecified
   </td>
   <td>Unspecified
   </td>
   <td>Specified
   </td>
   <td>Unspecified
   </td>
   <td>Aggregated container requests
   </td>
   <td>Infinity
   </td>
   <td>Specified
<p>
Container Request
   </td>
   <td>Infinity
   </td>
  </tr>
  <tr>
   <td>Unspecified
   </td>
   <td>Unspecified
   </td>
   <td>Unspecified
   </td>
   <td>Specified
   </td>
   <td>Aggregate of effective container requests
   </td>
   <td>Aggregate of effective container limits
   </td>
   <td>Effective Container Limits
   </td>
   <td>Specified Container Limits
   </td>
  </tr>
  <tr>
   <td>Unspecified
   </td>
   <td>Unspecified
   </td>
   <td>Unspecified
   </td>
   <td>Unspecified
   </td>
   <td>Aggregate of effective containers requests 
   </td>
   <td>Infinity
   </td>
   <td>Implementation defined minimum
   </td>
   <td>Infinity
   </td>
  </tr>
</table>
<br>

  Following validation logic needs to be added:

  1. Only CPU and memory resources are supported in pod level resource specifications, hence invalidate pod-level resource specification for other resources like gpu, storage, etc. 
  2. requests

>       * `pod.spec.resources.requests[memory/cpu] >= sum(pod.spec.containers[*].resources.requests[memory/cpu]) + sum(pod.spec.initContainers[*].resources.requests[memory/cpu])` is valid.

Reviewer: you don't need to sum the init containers, take a max of them, correct? Also account for sidecar containers. I would suggest to just define some function and use it everywhere, the function is quite complex with the sidecars

>       * `pod.spec.resources.requests[memory/cpu] >= sum(pod.spec.containers[*].resources.requests[memory/cpu]) + sum(pod.spec.initContainers[*].resources.requests[memory/cpu])` is valid.

Reviewer: as commented below, my understanding that this comparison is only executed if pod level requests are defined, correct? If not defined - there is nothing to compare with

Reviewer: if the assumption is that we will populate pod level requests with the effective value - please clarify. I didn't find the mention of it anywhere

                resources:
                  requests:
                    memory: 40Mi
                    cpu: 0.5
                  limits:
                    memory: 200Mi
                    cpu: 1 
          ```

          The pod-level limit will serve as an aggregate ceiling. This will cover the scenario where the sum of peak usage for individual containers in a pod is high, but these
          peaks don’t necessarily occur simultaneously.

          If we simply set the pod-level limit as the sum of the individual container limits, we might overallocate resources. The pod might never actually use all those resources
          at once, leading to wasted capacity on the node.

      * `pod.spec.resources.limits[memory/cpu] < pod.spec.containers[i].resources.limits[memory/cpu]` is invalid.

        A single container’s limit should not be allowed to exceed the pod level limit.
  4. Validation logic comparing requests and limits

      1. Pod level requests greater than pod level limits is invalid.

      2. Pod level requests greater than sum of limits of all containers is invalid.

      3. Sum of container level requests greater than Pod level limits is invalid.
        

>           This is already ensured by other validation rules in 1. and 2. as following:

Reviewer: those rules written in assumption that there is a pod level request defined. If pod level request is not defined, rules will not execute, so you need this another rule to check this condition.

        
          As per 1. Container level Requests <= Pod level requests is valid<br>
          And as per 3.1 Pod-level requests <= Pod level limits is valid<br>
          Hence, by transitivity these rules will ensure Container level requests <= Pod-level limits.

>       4. [Already exists] Sum of container level requests less than container level limits is invalid.

Reviewer: same comment as before - the note that it is already exist is misleading

#### Init Containers & Sidecar Containers

With the current implementation (before this KEP), Pod’s effective resource request/limit is the sum of 
pod overhead and the higher of:

> * The sum of all non-init containers(app and sidecar containers) request/limit for a resource.

Reviewer: the formulae is more complicated than this as sidecars are running while init containers are executed. You can link here: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/753-sidecar-containers/README.md#resources-calculation-for-scheduling-and-pod-admission

* The highest value resource request or limit among all the init containers.

```

> Effective Request/Limit = Max(Max(Init container limit), Sum(Sidecar containers) + Sum (App Containers)) + Pod Overhead

Reviewer: see comment above, not a correct formulae

  In this the pod level cgroup will be set using the requests and limits from the resources specified at pod-level.

  Example pod spec:

  ```yaml
  apiVersion: v1
  kind: Pod
  spec:
    resources:
      requests:
        memory: 100Mi
        cpu: 1
      limits:
        memory: 200Mi
        cpu: 2
    containers:
      - name: container-1
      - name: container-2
        resources:
          requests:
            memory: 50Mi
            cpu: 1
  ```

  **Resources requests at pod-level**

  * The scheduler will use the requested resources (CPU and memory) settings specified in the PodSpec to find a suitable node to place the pod. 
  * Kubelet will use the CPU requests from the pod-level resources section, and translate it to cpu.weight in pod-level cgroup.

  **Resources limits at pod-level**

  * All init(Regular and Sidecar) and non-init containers will be capped by the pod level limit which represents the peak usage of all containers.
  * Kubelet will use CPU and memory limits from PodSpec, and translate them to cpu.max and memory.max cgroup settings at pod-level respectively. 

  **Pros**

  * Simple to understand and implement.

>   * Guarantees enough resources for all containers.

Reviewer: this is a requirement in the current kubelet implementation. If we want to change this, I would suggest to run it as a separate KEP

  **Cons**

  * Can lead to over-provisioning (resource wastage), especially if the init container peak usage is significantly higher than the steady-state usage 
  of the other containers (sidecar and app containers). However this can be mitigated by setting container level resource spec for this case.
  
  Example 1: If your application containers require high resources simultaneously, set pod-level requests/limits such that it is an aggregate of all container-level requests/limits.

  Example 2: If your application has init containers (and not sidecar containers) with high resource requirements and usage, while your application containers require fairly less resources, 
  set pod-level requests/limits as maximum of init and non-init container resources requirements. 

  Example 3: For a Pod with init containers, sidecar containers and app containers, set the pod level requests/limits such that:
  
  ```
  Pod level Request/Limit = Max(Max(Init container limit), Sum(Sidecar containers) + Sum (App Containers))
  ```

> * **Option 2: Phased requests/limits - Pod-level cgroups are set after regular init containers are completed.**

Reviewer: if we do not do it, perhaps we can delete this section to avoid confusion. Or we are still undecided?

    * It is implicitly resourced (pod.spec.resources is not set) AND all containers requests are equal to their limits.

  **Note: If the pod spec includes `resources` stanza at pod level, but either requests or limits are missing for a resource, the pod will not be considered Guaranteed QoS, even if container-level requests and limits are present and are equal.**

* **Burstable QoS Criteria**

    A pod is considered `Burstable`, if for each resource type it doesn't meet the criteria for Guaranteed QoS for that resource, and either of following is true:

    * It is explicitly resourced (pod.spec.resources is set) for that resource, AND request (less than or equal to limit) or limit is set.
    * It is implicitly resourced for that resource AND for at least one container in the pod, request (less than or equal to limit) or limit is set. 
  

* **BestEffort QoS Criteria**

    None of pod-level or container-level resources requests or limits are specified.

#### OOM Killer Behavior
With Pod level resources, the OOM Killer will evolve to consider the cumulative resource limits set for the entire Pod. If the aggregated

> memory usage of all the containers within a Pod exceeds the effective pod level limits, the entire Pod becomes a candidate for OOM Killing, even if individual containers are within their limits.

Reviewer: Do you mean we will try to OOM kill the whole pod like we did for all processes inside the container: https://github.com/kubernetes/kubernetes/pull/117793?

Within a pod, the OOM Killer will maintain its existing behavior, considering container level limits (if set) and QoS classes, so a container
exceeding its individual limits can still be terminated even if a pod as a whole is under its limit.

#### Admission Controller

* LimitRanger

  LimitRanger today supports objects of type `Container`, `Pod` and `PersistentVolumeClaim`. It allows following constraints:
  
    * min (this option is to specify min usage constraint on the resource)
    * max (this option is to specify max usage constraints on the resource)
    * defaultRequest (this option is to specify default resource request for the named resource)
    * default (this option is to specify default limit for the named resource)
    * maxLimitRequestRatio (represents the max burst value of the named resource)
  

>   default and defaultRequest are not supported by LimitRanger of type `Pod`. To incorporate Pod level resource specifications, LimitRanger of type `Pod` logic will be extended with following changes:

Reviewer: what would be our long term recommendation in 5 years? Will we say - do defaulter and min/max for the Pod? Or keep it be on container level?

Reviewer: will we ever need per-container defaulter in 5 years?

Reviewer 3: I think users will likely have different preferences about setting pod vs container resources. Do you thin we need to be that prescriptive in this KEP?

Reviewer: There is a consistency problem we need to think about. Mostly from the perspective on what people are trying to express with the limit ranger. E.g. if min set for container, but the total pod resources set to the value, should limit ranger decline the pod? Another consistency question: https://github.com/kubernetes/enhancements/pull/4678#discussion_r1645196655

Besides the consistency issues above, thinking on how people will be using limit ranger in future may help define it's API now.

Reviewer 3: I'm still leaning towards not being overly prescriptive on these, or not making too many assumptions as long as there's a clear way for users to configure this as they want. 

@thockin for more input.

Author: I think we should let users use defaulters at both container and pod level and let them decide the right combination of values for defaults.

  1. Validation at Pod Admission will derive pod resources from `resources` stanza at pod level if it is set.

>   2. Extend `Pod` type LimitRanger to support `defaultRequest` and `default` to allow setting Pod level defaults.

Reviewer: will limit ranger have a consistency check? If container AND Pod defaulters are confiogured, limitranger can make Pod spec be incorrect (it is already can do it in some edge cases, I wonder if we increase the number of those edge cases)

Reviewer 3: Since limit ranger is implemented as a list, I think it's better to keep different limitRangers independent.

But we can look into better practices for beta

Author: yeah we can revisit this again for Beta to see if there's a change to reduce the number of edge cases.. But for now, we are complying with what's mentioned in the public doc: "A LimitRange does not check the consistency of the default values it applies." https://kubernetes.io/docs/concepts/policy/limit-range/#limitrange-and-admission-checks-for-pods

* ResourceQuota

>   Quota will derive the effective resource requests and limits for the pod, including the new pod-level resources.

Reviewer: interesting exercise for the rollout/rollback section on how quota calculation will be confused and what kubernetes will do when feature gate will be disabled. Is it covered?

#### [Scoped for Beta] CPU Manager

Note: This section includes only high level overview; Design details will be added in Beta stage.

With the introduction of pod-level resource specifications, the CPU manager in Kubernetes will adapt to manage CPU requests and limits at the pod level rather than solely at the container level. This change means that the CPU manager will allocate and enforce CPU resources based on the total requirements of the entire pod, allowing for more flexible and efficient CPU utilization across all containers within a pod. The CPU manager will need to ensure that the aggregate CPU usage of all containers in a pod does not exceed the pod-level limits.

#### [Scoped for Beta] In Place Pod Resize

> TBD. Do not review for the alpha stage.

Reviewer: will we allow inplace to go into beta without pod level support while the pod level support is in alpha? Will we consider in place pod resize to require pod level support when both features are in alpha? 

If not - we will need to rename the kep to "in place containers resize" to highlight that pods are out of scope

Reviewer 3: My thoughts: We shouldn't require a feature under development to support other alpha or even pre-alpha features. So in-place pod resizing going beta shouldn't be blocked on this feature. But better coordination/collaboration will help both.

I don't think we need to rename. We can just clearly call out the scope in the KEP. But @tallclair may have different opinions.

Reviewer: getting both of those features to GA in a current scope will result in inconsistent user experience for sure.

Also small issues like inplace will need to account for cases when changing container's resources will put a Pod in inconsistent state (e.g. sum of requests > pod's request) needs to be accounted while in alpha/beta.

Since inplace is already in alpha, I think the validation I mentioned must be added by this KEP. Worth mentioning in this document

Reviewer 3: > Also small issues like inplace will need to account for cases when changing container's resources will put a Pod in inconsistent state (e.g. sum of requests > pod's request) needs to be accounted while in alpha/beta.

In-place update simply provides a mechanism for users/components to update container resource specs. If the provided values are invalid, they'd be denied. We can further clarify the behavior in the KEP, but I think this is working as intended and is already in scope.

> #### [Scoped for Beta] VPA

Reviewer: so what will be behavior in alpha? If VPA is enabled, will it try to get Pod into the inconsistent state and fail to update it over and over again?

Reviewer 3: Ideally we can 1) make the API change first to support pod-level resource spec, while disallowing setting them and 2) update other components (e.g., VPA) to ignore pods with resource limit set until the support is in-place. 

In practice, VPA is an optional component where users configure the "targets" to scale. We can document that VPA is not supported yet in alpha and mention that enabling VPA could cause conflicts (e.g., VPA unable to update the resource request/limit) that renders scaling not work as expected.

Reviewer 6: ok to punt to beta phase.

TBD. Do not review for the alpha stage.

> #### [Scoped for Beta] Cluster Autoscaler

Reviewer: so what will be behavior in alpha? Will CA assume that pod doesn't need resorurces and will not create new nodes for it?

Reviewer 3: I think that's actually a desired behavior, but we need to check if that's true @ndixita

Reviewer: let's mention it than. If enabling of this feature while in alpha will cause CA failures, it needs to be at least mentioned.

Reviewer 3: @ndixita please add a sentence to describe the CA behavior when the alpha feature is enabled.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

<!--
This section must be completed when targeting alpha to a release.
-->

###### How can this feature be enabled / disabled in a live cluster?

<!--
Pick one of these and delete the rest.

Documentation is available on [feature gate lifecycle] and expectations, as
well as the [existing list] of feature gates.

[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
-->

- [X] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: PodLevelResources
  - Components depending on the feature gate: kubelet, kube-apiserver, kube-scheduler
  - Will enabling / disabling the feature require downtime of the control
    plane? No. `resources` must be set/unset in `spec` at pod-level to enable/disable this feature.
    The workloads need to restart to disable the feature.
  - Will enabling / disabling the feature require downtime or reprovisioning
    of a node? No. Only the workloads need to restart.

> ###### Does enabling the feature change any default behavior?

Reviewer: this needs to be mentioned: https://github.com/kubernetes/enhancements/pull/4678/files#r1645195855

Reviewer: this is still an open question here, correct? If LimitRanger was configured and somebody is trying this feature, they can work around limit rangers by setting the higher pod request than they allowed by limit ranger and quota since those will not support pod level resources initially. Not a big deal for alpha, just worth mentioning

Author: @jpbetz this question is for breaking change. Does any default behavior change if the feature is turned on, but not used deliberately...

Reviewer 9: I agree this is an interesting issue that should be tracked in post-GA criteria.

I don't consider it blocking for this PRR question, since the goal here is to identify if this enhancement is fully backward compatible with pre-existing system behaviors, and this is more of a question of how it integrates with other system behaviors when used (which is interesting and important, but not what PRR is concerned with for this question specifically)

Reviewer: I think handling this for beta (or at least documenting) is important. If a new feature allows to work around limits set by cluster administrator via LimitRanger, it creates a risk in rolling out the feature. So cluster administrators may need to get prepared for the rollout.

I would think it is important for beta (enabled by default) PRR.

No. This feature is guarded by a feature gate, and requires setting Pod level `resource` stanza explicitly.

> ###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Reviewer: the quota behavior needs to be explained here: https://github.com/kubernetes/enhancements/pull/4678/files#r1645197230

> ###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Reviewer: one fun aspect - if pods with pod limits were scheduled while feature gate was enabled, they will be scheduled on nodes with no resources.

Also CA may go highwire: https://github.com/kubernetes/enhancements/pull/4678/files#r1645206055

Reviewer 9: @SergeyKanzhelev Can you explain more? I'm not sure I follow, but want to make sure this is resolved before PRR approval.

Reviewer 3: Does this refer to pods with pod level resource requests (but no container level requests) may get scheduled onto nodes without sufficient resources when the feature gate is **disabled**? This seems to be an edge case that would only happen if the feature gets disabled AFTER the pod creation but BEFORE the pod scheduling.

Author: I think beyond the edge case, since CA doesn't take pod-level resources into account, if a user specifies only pod-level resources then CA will think that this pod doesn't need any resources as container-level values are not set which could match such pods to nodes with no resources. I'd say this is expected for now, until the support for CA comes in with Beta

Reviewer: yes, I meant that if somebody scheduled a bunch of deployments with pod level resources, and then feature was disabled, those Pods will be seen as not having any resources declared. And may cause some troubles in how they will be scheduled

> Yes. Once the feature gates are disabled, pods created with pod level resource spec need to be deleted and recreated with only container level resources.

Reviewer: we rarely have a requirement to delete all pods scheduled with this feature enabled.

If we require that - we need to delete those pods first and then disable the feature. It needs to be clearly stated in docs for alpha feature.

Reviewer 3: I don't think we need to mention the recreation. +1 to just disable the feature and delete resources using pod level resource spec.

Reviewer: deleting pods automatically is not expected on feature disablement.

It would be OK to say "Pods created with the feature on will retain the resources limits applied by the feature". Generally, we don't expect 
- automatically recreate the pods
- adjust running pods to disable the feature in those running pods.

We would expect that if the pod gets recreated by a deployment or something, that the new pod will not have the feature. But the act of redeploying should be a user decision. The behavior does need to be documented, though.

Some edge cases include - what would be the behavior of kubelet already allocated memory based on pod level resources and when the feature disabled a new Pod was scheduled on that node. Will it result in any issues? Those cases needs to be documented

Reviewer 3: See the reply in another comment: https://github.com/kubernetes/enhancements/pull/4678/files#r1645199565

We can document that there might be a mismatch in resource calculation and actual usage. To fix the issue, users can delete the running pods and create new ones.

That seems sufficient to address most of the issues IMO.

Reviewer 9: +1  to @SergeyKanzhelev here. What I'd like to see is that when the feature gate is disabled, no additional pods may use the feature, but pods that  already exist and are using the feature continue to work as expected.

###### What happens if we reenable the feature if it was previously rolled back?

> Rollbacks will require disabling the feature, and recreating the pods that were using this new feature.

Reviewer: it is rare that we require pods recreation on feature disablement. Also the question is asking what will happen with the running pods when feature was flipped back and forth

Reviewer 3: > Also the question is asking what will happen with the running pods when feature was flipped back and forth

The existing pods with the feature enabled should continue run without issues assuming they are already running. If they haven't been created on the node, kubelet would ignore the pod-level resource spec, and go ahead and set up the pods/containers with the container spec. In a way, this would be working as intended.

Reviewer 9: Note that we don't support rollback of alpha features.  So the question here is really:  "Once this feature goes to beta, and the previous version has the implementation and feature gate, what happens during rollback/reenablement?"   Ideally, the feature continues to work as expected when rolled back, but if the previous release did not have the feature gate enabled-by-default, no additional pods may be created with the feature, but once we reenable the feature both existing pods continue to work and new pods can be created that use the feature.

Once the feature is reenabled, the existing workloads will run as usual since they don't have `resources` set at pod level.
The new `resources` field will become available again for use in the Pod Spec for existing and new workloads.

###### Are there any tests for feature enablement/disablement?

Yes, the tests will be added along with alpha implementation. 
* Validate that scheduler considers pod level resources when the feature is enabled, and falls back to container level resources 
when the feature is disabled. 
* Validate the kubelet is using pod level resources in cgroup settings if the feature is enabled, and uses container level values
when feature is disabled.
* Confirm that the API server gracefully ignores the pod-level resource fields and reverts to the existing container-level 
resource aggregation logic when the feature gate is disabled.
* Confirm Limit Ranger uses defaultRequest and limit when feature is enabled, and errors out for these values when feature is disabled.

### Rollout, Upgrade and Rollback Planning

<!--
This section must be completed when targeting beta to a release.
-->

> ###### How can a rollout or rollback fail? Can it impact already running workloads?

Reviewer: this needs to be mentioned: https://github.com/kubernetes/enhancements/pull/4678/files#r1645195855

set resource constraints at pod level, limiting the flexibility and ease of
resource management for pods as a whole. This limitation is particularly
problematic when users want to focus on  controlling the overall resource
consumption of a pod without the need to meticulously configure resource
specifications for each individual container in it.

To address this, this KEP proposes extending the Pod API to support Pod-level
resource limits and requests for non-extended resources (CPU and Memory), in
addition to the existing container-level settings.

## Motivation

Kubernetes workloads often consist of multiple containers collaborating within a
pod. While this separation enables granular monitoring and management of each
container, it can be challenging to accurately estimate and allocate resources
for individual containers, especially for workloads with unpredictable or bursty
resource demands. This often leads to over-allocation of resources to ensure that
no container experiences resource starvation, as kubernetes currently lacks a
mechanism for sharing resources between containers easily. 

Introducing pod-level resource requests and limits offers a more simplified

> resource management for multi-container pods as it is easier to gauge theß

Reviewer 8: ```suggestion
resource management for multi-container pods as it is easier to gauge the
```

1. This KEP focuses on the core resource types of CPU and memory. It doesn’t
intend to address other resource types (e.g. GPU, storage, network bandwidth) at
the pod level in this phase. However, it could be considered in future extensions of
the KEP.

## Proposal

### User Stories

#### Simplified Resource Management

There are cases when it is challenging to decide a-priori how much resources to
allocate to each container in a pod. For instance, consider a development
environment implemented as a Kubernetes Pod with multiple sidecar containers for
tools, and a regular container for web-based IDE. The IDE container may have a
fixed memory limit, but the resource needs of the tool containers can fluctuate
depending on the developer's activities.

The proposed Pod Level Resource Spec feature addresses this by allowing all
containers within the pod to share a defined resource pool. In the example
provided, even though the IDE has its own resource constraints, tool1 and tool 2
containers can dynamically utilize any remaining resources within the pod's
overall limit. 

```yaml
apiVersion: v1
kind: Pod
metadata:
  labels:
    run: ide
  name: myide
spec:
  resources:
    limits:

>       memory: "1024Mi"

Reviewer: is the mix of `Mi` and `M` intentional in this pod spec?

      cpu: "4"
  initContainers:
    - image: shell
      name: debian:buster
    - image: tool1
      name: first-tool:latest
    - image: tool2
      name: second-tool:latest
  containers:
    - name: ide
      image: theia:latest
      resources:
        requests:
          memory: 128M
          cpu: "0.5"
        limits:
          memory: 256M
          cpu: "1"
```

> Without this feature, the developers will need to estimate beforehand how much

Reviewer: the pod spec above shows 3 init containers that will be completed by the time `ide` will start. Was it meant to be sidecars?

In scenarios where multiple containers within a pod experience independent and
uncoordinated resource usage spikes, the current container-level resource
management can lead to significant resource wastage. This is because the
resources are allocated based on each container's peak resource usage, even though
the peaks might not occur simultaneously. Consequently, while one container is
experiencing high resource usage, the other containers might be idle, resulting
in underutilized resources within the pod.

Pod Level Resources offer a solution for a more efficient resource allocation
strategy. By setting a limit at pod level, resources can be shared dynamically
among containers. This means that when one container experiences a spike, it can
utilize the idle resources of the other containers within a pod. 

Since resource bursts across containers are not synchronized, the aggregate
resource usage of the pod, in this case, is generally lower than the sum of the
individual container's peak usage. By setting the pod level limit to a percentile
of this aggregate usage, we can achieve better resource utilization and avoid
unnecessary over-allocation.

### Notes/Constraints/Caveats

> 1. cgroupv1 has been moved to maintenance mode since Kubernetes version 1.31.

Reviewer: do we plan to add an API-level validation for this somehow? Or just fail the pod admission on the kubelet?

   Hence this feature will be supported only on cgroupv2.
2. Pod Level Resource Specifications is an **opt-in** feature, and will have no
effect on existing deployments. Only deployments that explicitly require this
functionality should turn it on by setting the relevant resource section at
pod-level in the Pod specification.

### Risks and Mitigations

1. Since Pod Level Resource Specifications is an **opt-in** feature, merging the

>    feature related changes won't impact existing workloads. Moreover, the feature

Reviewer: same point as  below. If there are thrid party schedulers or other tools assuming how resources are being calculated, they may break. We have/had the same risk with the sidecar containers. Worth mentioning

   This means that any workloads or tools that depend on reading or interpreting
   container cgroup values might observe different behavior if they start using
   pod level resource spec without specifying container level settings. This
   aspect of the feature will be clearly highlighted in the feature documentation
   to make users aware of this change and the potential impact on their workloads.
   This way, they can adjust the pod level and container level settings
   accordingly as needed.

## Design Details

### Components/Features changes

#### Cgroup Structure Remains unchanged

This KEP introduces pod-level resource specifications but does not alter the
fundamental cgroup structure. Currently, pod level cgroups are derived from the
resource requests and limits defined at the container level in the PodSpec. With
this KEP, the PodSpec is extended to include pod-level resource requests and
limits. These values will be directly used to configure the pod-level cgroup,
providing a more direct and explicit way to manage the overall resource
consumption of the pod.

#### PodSpec API changes

New field in `PodSpec`

```
type PodSpec struct {
...

>   // Compute resource requirements.

Reviewer: the field description will need to be expanded to describe the feature gate needed, the lack of cgroupv1 support, the limitation that only cpu and memory are currently supported and a relationship with other resource definitions (maybe in a form of a link)

  // +optional
  Resources ResourceRequirements
}
```

#### PodSpec Validation Rules

##### Proposed Validation Rules
* **Symmetric Semantic**: Just as container-level requests default to
  container-level limits when container-level requests are unspecified, the
  proposal extends this logic to the pod level: if the pod-level request is

>   not specified, it defaults to the pod level limit.

Reviewer 8: This makes sense when container requests aren't specified, but I think it could be surprising in the case when they are. Consider this case:

- I have a Burstable pod with a couple containers that specify requests << limits
- Although each container can burst up, I want to cap the simultaneous burst limit, so I add a pod-level limit
- Since I didn't set a pod-level request, the pod request is defaulted to the limit, and the pod is now guaranteed

This seems like it would be a relatively common use-case, and to support it with the requests=limits defaulting, I would need to manually total the container requests (which can be complicated with init containers & sidecars).

Reviewer 8: My expectation for how this would work is:
- If container requests are set, then the pod request is the sum of the container requests (meaning that adding a pod limit to an existing pod wouldn't change the requests)
    - If `sum(container requests) > pod limits`, then fail validation
- If container requests are not set, but pod limit is set, then pod request == pod limit

Author: @tallclair IMO it gets complicated in the mix containers scenario: one container has request set, the second one doesn't. What do we do in this case? Should we derive the pod level request from container level requests only when request is set for all containers? 

We could derive pod level requests from container level requests if it is set for all containers only. But it does lead to different behavior in different scenarios which requires more thinking about how the values are derived. We rather wanted to make this simple to use.

If, for a user, it is complicated to derive pod level request from container level requests (although they know the requests for all their containers), one could argue that deriving pod level limit is also challenging but the user did specify the limit at pod level. So they should be able to estimate the minimum resource guarantees they would need for the overall pod level?

Author: Trying to think about this again with the following example:

Consider a pod with Pod Limit = 200 Gi; init container request = 50Gi, regular container with no request and limit. 
KEP proposed approach:
Derived Pod Request = 200 Gi; Derived init container limit = 200 Gi, Derived regular container request = min, Derived regular container limit = 200 Gi.

We will document that this is how the defaults would be applied. If you, as a user, want a different memory guarantee at pod level then please specify it. 

@tallclair could you please help me understand what the derived values will look like with the approach you are suggesting?

* **Pod-Level Limits Enforce Strict Resource Constraints:** When pod-level limits
  are set, they act as a hard cap for the total resource consumption of all the 
  containers within the pod.
* **Sum of Container Level Requests is Never Greater Than Pod Level Request**:
  The combined resource requests of all the containers within a pod will always
  be less than or equal to the defined or defaulted pod-level request.
* **Container Request if unspecified is set to implementation defined minimum**:
  If a container's resource request is not specified, but pod level requests
  and/or limits are set, the container request will be an implementation defined

>   minimum value. 

Reviewer: what effect will it have on CPU shares? This comment is not a concern, just thinking out loud if this defaulting to minimum may lead to sum unfairness when calculating cpu shares for container cgroups

Reviewer 12: +1
I'm not sure what's the use-case for setting pod-level CPU requests if the `cpu.weight` won't affect any of the containers. The only thing I can think of is that it will decrease the density of pods on the node.

The table below presents a comprehensive view of how kubernetes will
determine the applied (or effective) resource requests and limits at both the pod
and container level, considering various combinations of user-specified pod-level
and container-level settings. Rows 1 - 4 states the existing rules that are
applied for different combinations of the container level and limits. Rows 5 - 16
outlines the proposed rules with the addition of pod-level resource
specifications.

|Sr.No| **Pod Level <br> Requests** | **Pod Level <br> Limits**  | **Container Level <br>Requests** | **Container Level <br>Limits**  | **Applied Pod <br>Level Requests** | **Applied Pod <br>  Level Limits**   | **Applied Container <br>  Level Requests** | **Applied Container <br>  Level Limits**   |
|-----|--------------|-------------|--------------|-------------|--------------|--------------|--------------|--------------|
| 1.  | Unspecified  | Unspecified | Unspecified  | Unspecified | &sum;{CR}    | max    | Implementation defined min | max  |
| 2.  | Unspecified  | Unspecified | Unspecified  | Specified   | &sum;{CL}    | &sum;{CL}    | CL           | Specified CL |  

> | 3.  | Unspecified  | Unspecified | Specifiec    | Unspecified | &sum;{CR}    | max          | Specified CR | max          |

Reviewer 8: ```suggestion
| 3.  | Unspecified  | Unspecified | Specified    | Unspecified | &sum;{CR}    | max          | Specified CR | max          |
```

   * Defaulting container requests to container limits: If container requests are
     unspecified, set them equal to container limits (which were just set to the
     pod limit).
   * Aggregating Container Requests: Calculate the pod level request by adding up
     the requests of all the containers.
   * Equalizing container limits with pod limit: When container level limits are
     not set, each container's resource limit is set to match the pod's overall
     limit.   

  This approach can lead a situation where the calculated pod request (derived by
  aggregating container requests) exceeds the defined-pod limit, which is 
  logically inconsistent. Furthermore, this approach requires a multi-step
  calculation to derive pod request that might be less intuitive to grasp. It
  involes first setting container limits, then defaulting the container requests
  to those limits, and finally aggregating those requests to determine pod
  request.
  
  The examples in the next section will further illustrate the issue and how the
  chosen approach (Option 2) provides a simpler alternative.

  **Option 2 (Chosen): Set Pod Level Request equal to Pod Level Limit for Simplicity.**
 

>    * Defaulting pod request to pod limit: If the user doesn't explicitly specify

Reviewer: so if pod level is set to 100, when there are two containers each asking for 60 request and limit, we are entering the "bad state" by defaulting pod level request to limit?

It is similar to @tallclair comment above I think

Author: This will be an invalid pod spec. There's a validation rule this kep proposes: "Sum of Container Level Requests is Never Greater Than Pod Level Request".

Reviewer: hm, yes, I was thinking of something else. If total request is larger than the limit of the Pod - it does makes sense to make this pod invalid. I will think more about what confused me

     a pod level request, it is automatically set to be equal to the defined pod
     level limit. By mirroring the existing behavior at container level, where
     unspecified container requests default to their respective container limits,
     this KEP extends the same logic to pod level resource management, where
     unspecified pod requests default to pod limits. 
   * Container Request adjustments: When container-level request is not set, it
     is set to an implementation defined minimal value.

>    * Equalizing container limits with pod limit: When container level limits are

Reviewer: I assume this includes init containers, correct?

    memory.max cgroup settings at pod-level respectively. 

  **Pros**

  * This approach is straightforward to understand and implement.
  * It ensures enough resources are available for all containers.
    within the pod, collectively
  
  **Cons**

  * This method can sometimes lead to potential over-provisioning (resource wastage),
    particularly when the init containers have significantly higher peak resource
    usage compared to the non-init containers. In this case, allocating resources
    based on the init container's needs can result in unnecessary resource
    allocation for the application containers. However this can be mitigated by
    setting container level resource spec for application containers.

    For example, if your init container needs 2 CPU cores and 512Mi of memory
    for a short burst during initialization, while your application container
    only requires 0.5 CPU cores and 256Mi of memory. You should set pod level
    requests and limits to 2 CPU and 512Mi of memory. Crucially, you could also
    set container level requests and limits for your app container (0.5 CPU and
    256Mi). This prevents the application container from using the extra

>     resources allocated to the pod, making the resources available for other workloads.

Reviewer 8: This raises an interesting question for resize. A nice solution to this over-provisioning problem is to just resize the pod after it's initialized. But what if I want to resize the pod requests to be lower than the init container requests? Should we say that the init container is completed, so we disregard those requests? Or do we require the init container to be resized too? What if the node restarts, and the init containers are re-run? Food for thought...

Reviewer: our guarantee today is that we can reinit containers at any moment. If we want to change it, we need to discuss

prioritize those values when determining the Quality of Service (QoS) for a pod.
Here's the order of preference:
1. Pod Level Resources: If the resources are specified at the pod level,
   Kubernetes will use those values to determine the QoS class.
2. Container Level Resources: If pod-level resources are not specified,
   Kubernetes will back to the resources defined at the container level for QoS
   classficiation. The value of resources specified in the container’s
  
Essentially, pod-level resource specifications, when present, take precedence
over container-level specifications in determining the QoS class of a pod.

For each core resource type (CPU, Memory), a pod may have pod-scoped resource
request and/or limit. Pods which have a defined request or limit for a given
resource are said to "have explicitly resourced" or "be explicitly resourced".
Pods which DO NOT have a defined request or limit for a given resource type are
said to "have implicit resources" or "be implicitly resourced"

* **Guaranteed QoS**

    A pod is considered `Guaranteed` if, for each resource type (CPU and memory),
    one of following conditions is true:

>     * It is explicitly resourced for that resource type AND the the pod's request

Reviewer: one interesting side effect from this will be that the definition of "Guaranteed" will change a bit. Today guaranteeed indicates both - kuberentes side and app side. K8s side: there is always as much resources as you requested and there are no noisy neighbours who can cause your Pod to be evicted. App side: if app uses exact resources specified as a limit on a cgroup - it will always run fine. With the change - we are keeping the k8s side of it regarding noisy neighbours. But the app side of it is not true any longer. Containers will need to coordinate usage to not step into each other's limits.

Reviewer 6: I believe we discussed this at SIG Node design review a while back. I disagree with the comment. 

Please note that pod-level resources are optional. This means the existing behavior of Guaranteed QoS with container-level guarantees remains fully intact. The new feature extends the capabilities, offering more flexibility without taking away existing functionalities.

Reviewer 3: I think this is actually quite intuitive:
 - If you specify resources at the container level, it's guaranteed at the container level
 - If you specify resources at the pod level, it's guaranteed at the pod level
 
And like Dawn mentioned, it doesn't change the semantics today. If we want to evolve the QoS scope and definition in the future, this (container vs pod level) may come into play. We'll just need to be mindful of it.

skew implementation.

For users wanting to use this feature, it is user's responsibility to use the correct version of kube-scheduler,
kube-apiserver and kubelet otherwise the feature will not work as expected.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

<!--
This section must be completed when targeting alpha to a release.
-->

###### How can this feature be enabled / disabled in a live cluster?

<!--
Pick one of these and delete the rest.

Documentation is available on [feature gate lifecycle] and expectations, as
well as the [existing list] of feature gates.

[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
-->

- [X] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: PodLevelResources
  - Components depending on the feature gate: kubelet, kube-apiserver, kube-scheduler
  - Will enabling / disabling the feature require downtime of the control
    plane? No. `resources` must be set/unset in `spec` at pod-level to enable/disable this feature.

>     The workloads need to restart to disable the feature.

Reviewer 9: The question we're really looking to have answered here is "Will enabling / disabling the feature gate require downtime of the control plane?"

It's normal for pods that have the spec field set continue to use the feature gate even after the feature gate is turned off, so that part seems fine.

  - Will enabling / disabling the feature require downtime or reprovisioning

>     of a node? No. Only the workloads need to restart.

Reviewer 9: Same as above.

###### Does enabling the feature change any default behavior?

No. This feature is guarded by a feature gate, and requires setting Pod level `resource` stanza explicitly.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes. Once the feature gates are disabled, pods created with pod level resource spec need to be deleted and recreated with only container level resources.

###### What happens if we reenable the feature if it was previously rolled back?

Rollbacks will require disabling the feature, and recreating the pods that were using this new feature.
Once the feature is reenabled, the existing workloads will run as usual since they don't have `resources` set at pod level.
The new `resources` field will become available again for use in the Pod Spec for existing and new workloads.

###### Are there any tests for feature enablement/disablement?

Yes, the tests will be added along with alpha implementation. 
* Validate that scheduler considers pod level resources when the feature is enabled, and falls back to container level resources 
when the feature is disabled. 

> * Validate the kubelet is using pod level resources in cgroup settings if the feature is enabled, and uses container level values

Reviewer 9: Also we need to make sure that pods using the feature can only be created when the feature gate is enabled.

no container experiences resource starvation, as kubernetes currently lacks a
mechanism for sharing resources between containers easily. 

Introducing pod-level resource requests and limits offers a more simplified
resource management for multi-container pods as it is easier to gauge theß
collective resource usage of all containers in a pod rather than predicting each
container’s individual needs. This proposal seeks to support pod-level resource
management, enabling Kubernetes to control the total resource consumption of the
pod, relieving users from the burden of meticulously configuring resources for
each container. This simplified approach is particularly valuable for tightly
coupled applications, and it will enable Kubernetes to leverage the underlying
cgroup resource management mechanisms more effectively for better resource
utilization for bursty workloads. 

This new feature of resource specification at pod-level would complement, not
replace, the existing container-level limits, providing users with greater
flexibility.

### Goals

This proposal aims to:

1. Extend Pod API to allow specifying resource spec (requests and limits) at the
   pod level. 

> 2. Make this feature compatible with existing usage of container level

Reviewer 5: Worth calling out VPA in scope?

   requests/limits, and other features like topology manager, memory manager,
   etc.

> ### Non Goals

Reviewer 6: Can we explicitly add the following items into Non Goals: 

1) Pod-level device plugins: existing container-level device plugins are compatible with this proposal, but there are some discussion on how to share device(s) among containers within a pod. Treat those as none goal explicitly.
2) Dynamic QoS class adjustments: As we discussed offline, we agreed a static QoS class determination based on pod-level and container-level resources. Please explictly state that exploring dynamic QoS changes based on runtime conditions or pod phases is Non Goal. We can explore the possibility in the future. 
3) Fine-grained control over resource sharing within a pod: While pod-level resources enable basic resource sharing, more advanced controls (e.g., weighted sharing, prioritization) could be considered as a future enhancement.
4) Performance optimizations for specific workloads: While the feature is expected to improve resource utilization, specific performance optimizations for certain workloads (e.g., high-performance computing) can be explored in future iterations.

1. This KEP focuses on the core resource types of CPU and memory. It doesn’t
intend to address other resource types (e.g. GPU, storage, network bandwidth) at
the pod level in this phase. However, it could be considered in future extensions of
the KEP.

## Proposal

### User Stories

#### Simplified Resource Management

There are cases when it is challenging to decide a-priori how much resources to
allocate to each container in a pod. For instance, consider a **development
environment** implemented as a Kubernetes Pod with multiple sidecar containers for
tools, and a regular container for web-based IDE. The IDE container may have a
fixed memory limit, but the resource needs of the tool containers can fluctuate
depending on the developer's activities.

The proposed Pod Level Resource Spec feature addresses this by allowing all
containers within the pod to share a defined resource pool. In the example

> provided, even though the IDE has its own resource constraints, tool1 and tool 2

Reviewer 5: ```suggestion
provided, even though the IDE has its own resource constraints, tool1 and tool2
```

2. Pod Level Resource Specifications is an **opt-in** feature, and will have no
effect on existing deployments. Only deployments that explicitly require this
functionality should turn it on by setting the relevant resource section at
pod-level in the Pod specification.

### Risks and Mitigations

1. Since Pod Level Resource Specifications is an **opt-in** feature, merging the
   feature related changes won't impact existing workloads. Moreover, the feature
   will be rolled out gradually, beginning with an alpha release for testing and
   gathering feedback. This will be followed by beta and GA releases as the
   feature matures and potential problems and improvements are addressed.

2. While this feature doesn't alter the existing cgroups structure, it does
   change how pod-level cgroup values are determined. Currently, Kubernetes
   derived these values from the container-level cgroup settings. However, with
   Pod Level Resource Specifications enabled, pod-level cgroup settings will be
   directly set based on the values specified in the pod's resource spec stanza,
   if set. This change in behavior could potentially affect:

>     * Workloads or tools that rely on reading cgroup values: This means that any

Reviewer 5: We could look into projecting pod level values as well into the container either as a read only fs or through downward API.

Author: For the runtimes relying on container-level values, do they rely only on limits? If yes, the KEP talks about setting the pod-level limits in cgroups. Will that suffice?

Reviewer 5: Yes, that should work :+1:

   this change and its potential impact. This will allow users to:
    * Adjust their pod-level and container-level resource settings as needed
    * Modify any custom schedulers or tools to align with new resource
      calculation method.

## Design Details

### Components/Features changes

#### Cgroup Structure Remains unchanged

This KEP introduces pod-level resource specifications but does not alter the
fundamental cgroup structure. Currently, pod level cgroups are derived from the
resource requests and limits defined at the container level in the PodSpec. With
this KEP, the PodSpec is extended to include pod-level resource requests and
limits. These values will be directly used to configure the pod-level cgroup,
providing a more direct and explicit way to manage the overall resource
consumption of the pod.

#### PodSpec API changes

New field in `PodSpec`

```
type PodSpec struct {
...
  // The total amount of compute resources required by all containers in the pod.
  // Currently, only CPU and Memory resources are supported .
  // +featureGate=PodLevelResources
  // +optional
  Resources ResourceRequirements
}
```

#### PodSpec Validation Rules

> ##### Proposed Validation Rules

Reviewer 13: I'm unsure if it's specified anywhere in this KEP, but I couldn't find it.
What happens if multiple containers specify their resources, and the sum of those resources is greater than the Pod resources specified?

Reviewer 13: I found it:
>  **Rule: Aggregated container requests cannot be greater than pod-level
  request or limit.**
> * Aggregate container-level limits can be greater than pod-level limits, but the
  combined resource consumption of all the containers within a pod will always be
  less than the limit.

    * Derived From Containers: If no pod-level limit is defined, Kubernetes can
      derive it from container-level requests. Users should ensure that at least
      one container, including init containers, regular containers, or sidecars
      (including those injected later), has a request specified for this work
      effectively.
    * Combined Approach: Users can specify both container-level and pod-level
      requests. This can be useful for fine-tuning the priority order in OOM
      killing scenarios, as the oom_score_adj calculation will take both
      pod-level and container-level requests into account. (More details about
      oom_score_adj in #oom-score-adjustment). The scheduler will however use
      pod-level request, in this case, to find a suitable node.

  This ensures that Kubernetes can determine a minimum resource requirement for
  the pod, either directly or indirectly from the container requests (using the
  formula in [KEP
  #753](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/753-sidecar-containers/READMEmd#resources-calculation-for-scheduling-and-pod-admission))
  

>   **Rule: If no requests are provided either at the pod level or at the container level,

Reviewer 5: This sounds like a breaking change to the existing best effort behavior assuming no pod level requests are set.

Author: Correct, I will discuss and change this. Today we assume arbitrary container level minimum values. Do you have suggestions for what the minimum at pod level could be? No. of containers * the minimum values that we have today? 

These arbitrary values that we have today, seem very high for mininum resource requests… and they are used only by the scheduler to find a suitable node.

https://github.com/kubernetes/kubernetes/blob/60efdae3d1c4f5ac7e3aa9acbfa8a818e1fb2093/pkg/scheduler/util/pod_resources.go#L36

    default to the allocatable resources of the node where it is scheduled. This
    ensures the pod can utilize the maximum available resources on the node.

  **Rule: If Pod-level limit is not set, or if it cannot be derived from
  container-level limits (not all containers have limits set), the containers in
  the pod can use all the resources allocatable on the node.**
* The aggregated resource requests of all containers within a pod will never
  exceed the pod-level resource requests or limits.

  **Rule: Aggregated container requests cannot be greater than pod-level
  request or limit.**
* Aggregate container-level limits can be greater than pod-level limits, but the
  combined resource consumption of all the containers within a pod will always be
  less than the limit.

  **Rule: Aggregated container limits can be greater than pod-level
  request or limit, but individual container limit greater than pod-level limit
  is not allowed.**
 
* **Supported Resources**: Currently, pod level resources support only CPU and
  memory.

  **Rule: Pod-level resource spec for resources other than CPU and memory is invalid**

> ##### Comprehensive Tabular View

Reviewer 6: Can you explicitly include the design principle we discussed offline here, so that the readers understand the rationale behind. I noticed that you already include most of them below but scattered throughout the document. Centralizing them in a dedicated section would significantly enhance clarity and readability.

1) Prioritize explicit pod-level settings to respect the user's explicit intent and provides clarity
2) Avoid implicitly applying default values at the pod level. 
3) Consistent derivation from container values to ensure consistency and predictability
4) The derivation logic maintains the compatibility with existing behavior

There are several principles we followed during the design, but I failed to spot them from the doc. Maybe we can explicitly state them too:

1) Fairness: The resource allocation mechanism should strive for fairness among containers within a pod, preventing any single container from monopolizing resources unless explicitly configured. For example, OOM score adjustment calculation below demonstrates the fairness in our design. 
2) Flexibility: We carefully consider the flexibility in resource allocation, enabling users to express different resource requirements and sharing strategies as needed. For example, optional  pod-level resources, combination of pod-level and container-level limits, etc. make the resource allocation flexible enough to meet the different requirements.

[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
-->

- [X] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: PodLevelResources
  - Components depending on the feature gate: kubelet, kube-apiserver, kube-scheduler
  - Will enabling / disabling the feature require downtime of the control
    plane? No. Once the feature is disabled, the control plane components will
    stop using the pod-level resources from the spec.
  - Will enabling / disabling the feature require downtime or reprovisioning
    of a node? No. Once the feature is disabled, the kubelet will stop using the pod
    level resources from the spec.

###### Does enabling the feature change any default behavior?

No. This feature is guarded by a feature gate, and requires setting Pod level `resource` stanza explicitly.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes. Any new pods created after disabling the feature will not utilize the
pod-level resources feature. Disabling the feature will not affect existing pods
-- they will continue to function. 

> However, tfor pods created with pod-level resources, there will be a mismatch

Reviewer 9: ```suggestion
However, for pods created with pod-level resources, there will be a mismatch
```

between the resource calculations and the actual usage. To fix thie issue, users
may delete the running pods and create new ones.

###### What happens if we reenable the feature if it was previously rolled back?

Reenabling the feature will not affect existing pods. Once reenabled, any new

> pods can be created with pod-level resources to use the feature.

Reviewer 9: Can we say any more about what happens to a pod that used pod-level resources before the the feature was disabled?  I'm not quite sure what "will not effect'  implies here.  Would such pods have the pod-level resource fields set but no honored?  If so should we recommend the pods be recreated again after reenabling the feature?

     As both containers in the pod are regular containers, the pod-level request
     will be calculated as sum of container level request which is equal to
     100Gi.
    
   * Container-level Limit: Container-level limit will be set equal to
     pod-level limit only in cgroup settings (and not in PodSpec) for runtimes (or other softwares)
     relying on container-level cgroup values.

4. Pod-level Limit is set; Container Requests & Limits are set for 1 regular container.
   
   ![](./example-4.png)
   
   * Pod-level requests will be derived from container-level requests as at least
     1 container has requests set. The pod-level request, in this case is equal
     to 50Gi.

   * Container-level Limit: Container-level limit for container 2 will be set equal to
     pod-level limit only in cgroup settings (and not in PodSpec) for runtimes (or other softwares)
     relying on container-level cgroup values.    

5. Pod-level request is set; Container level limits set for 1 container.

    ![](./example-5.png)

   * As container-level limits are not set for all containers (only 1 container
     has limits specified) in this pod, there's no way to derive pod-level

>      limits. Hence the containers in this pod can use all the memory up to node

Reviewer 5: Does it make sense to instead require pod level limit be specified in this case?

        |-----------------|-----------|-----------|       
        | **Container 1** |   950     |   940     |
        | **Container 2** |   900     |   890     |
        | **Container 3** |   1000    |   990     |

    The new formula assigns lower oom_score_adj values to all containers in Pod 2
    than the containers in Pod 1, making them less likely to be killed compared
    to Pod 1. The reasoning behind this intent is to **reward the pods with
    explicit pod-level requests**.

    Also, this formula ensures that **containers with explicit requests** are
    prioritized over the containers with no requests set. In our example, within Pod 2, containers
    with explicit requests (Containers 1 & 2) have lower oom_score_adj values
    than Container 3 (no explicit request).
      

##### System Out-of-Memory (OOM) Kill

  When a node is under memory pressure, if the kernel is not able to reclaim
  memory, the kernel will trigger OOM Killer. The OOM Killer prioritizes killing
  the processes  with higher OOM scores. The new oom_score_adj formula, which
  considers both pod-level and container-level resources, helps the OOM Killer

>   makes fair and informed decisions about which processes to terminate.

Reviewer 5: ```suggestion
  make fair and informed decisions about which processes to terminate.
```

  
    * min (this option is to specify min usage constraint on the resource)
    * max (this option is to specify max usage constraints on the resource)
    * defaultRequest (this option is to specify default resource request for the named resource)
    * default (this option is to specify default limit for the named resource)
    * maxLimitRequestRatio (represents the max burst value of the named resource)
  
  default and defaultRequest are not supported by LimitRanger of type `Pod`. To incorporate Pod level resource specifications, LimitRanger of type `Pod` logic will be extended with following changes:
  1. Validation at Pod Admission will derive pod resources from `resources` stanza at pod level if it is set.
  2. Extend `Pod` type LimitRanger to support `defaultRequest` and `default` to allow setting Pod level defaults.

* ResourceQuota

  Quota will derive the effective resource requests and limits for the pod, including the new pod-level resources.

#### Eviction Manager

For memory.available signal, the eviction manager checks the pod's memory usage and compares it with memory requests at pod level to determine the pod eviction order. It aggregates the 

> container-level memory requests to calculate the effective pod-level memory request.When pod-level memory requests are specified, the Eviction Manager should use the pod-level request 

Reviewer 5: ```suggestion
container-level memory requests to calculate the effective pod-level memory request. When pod-level memory requests are specified, the Eviction Manager should use the pod-level request 
```

        specified but a limit it set, the Pod Spec is modified to set the [request
        equal to limit](https://github.com/kubernetes/kubernetes/blob/005f184ab631e52195ed6d129969ff3914d51c98/pkg/apis/core/v1/defaults.go#L167-L228) for each container. This modified Pod spec is then used by
        the scheduler, and also for configuring cgroup settings.  

  * **Rows 13 - 16**

    ![](./pod-request-limit-set.png)

      * When both pod-level request and limit are explicitly specified,the scheduler and
        cgroups will use those values directly.
    
      * Rows 13, 15 Annotation [c]: When container-level limit is not set, the
        pod-level limit is applied to each container's cgroup maximum value. This
        is because container-level limit is implied in this case, and some
        runtimes (like the Java runtime) rely on container-level cgroup maximum
        values for fine tuning their components.
<br>

##### Clarifying Tricky Cases and Design Decisions with Pod Level Resources

1. [Invalid Pod Spec] Row 5 from the scenarios table: Pod Limits Defined; Pod
   Requests & Container Specs Unspecified.
   
   ![](./example-1.png)
  

>    * The Challenge of Missing Requests: In this case, Kubernetes cannot determine

Reviewer 4: Given all we discussed today, would this not set PR from PL?

At L318 you said:

 "This means that at the pod level, if a resource request is not explicitly
   defined, Kubernetes attempts to derive it from requests specified
   at container-level within the pod. However, container-level requests are not
   set for at least 1 container, the pod-level request cannot be
   accurately derived. In this case,  the pod's resource request will default to
   the pod's limit, if pod limit is set."

Author: I think the pull request was updated after your comment or I was in middle of making that change. This is update now. Need to proof-read the entire KEP for missing details.

     the appropriate pod-level request as pod-level request is not set explicitly,
     and since container-level requests are also not set, there's no way for
     kubernetes to derive pod-level requests from the resource spec.
   * Prioritizing Explicit Resource Requests: Instead of relying on potentially
     inaccurate default values or guessing the user's intent, this KEP proposes
     rejecting such a pod specification. This encourages users to explicitly
     define their resource requests, promoting clarity and preventing ambiguity in
     resource allocation.  

2. [Invalid Pod Spec] Row 6 from the scenarios table: Pod Limits and Container
   Limits Defined, Pod Requests and Container Requests Unspecified.

   ![](./example-2.png)

>    * This pod will be rejected for same reasons as example 1 above (for row 5).

Reviewer 4: I think this one has a far more obvious reason to fail:

* The Container requests are defaulted from limit (60, 60)
* The aggregate request (120) is greater than the pod's limit

Author: Correct

     This KEP prioritizes specifying explicit resource requests at pod-level or at
     container-level.

3. Row 7 from the scenarios table: Pod Limit is set; Container Requests are set for
   both regular containers.

   ![](./example-3.png)

   * Pod-level Request: Since container-level requests are set, Kubernetes can
     determine the pod-level request using the formula from
     [KEP#753](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/753-sidecar-containers/README.md#resources-calculation-for-scheduling-and-pod-admission).
     As both containers in the pod are regular containers, the pod-level request
     will be calculated as sum of container level request which is equal to
     100Gi.
    
   * Container-level Limit: Container-level limit will be set equal to
     pod-level limit only in cgroup settings (and not in PodSpec) for runtimes (or other softwares)
     relying on container-level cgroup values.

> 4. Pod-level Limit is set; Container Requests & Limits are set for 1 regular container.

Reviewer 4: I'm not arguing to re-adjudicate at this time, but I wanted to point out that we are OK with assuming request=0 here, but we are not OK assuming request=0 when no container requests are set.

Author: PTAL, I updated finished updating all  the examples now.

the pod level in this phase. However, it could be considered in future extensions of
the KEP.

## Proposal

### User Stories

#### Simplified Resource Management

There are cases when it is challenging to decide a-priori how much resources to
allocate to each container in a pod. For instance, consider a **development
environment** implemented as a Kubernetes Pod with multiple sidecar containers for
tools, and a regular container for web-based IDE. The IDE container may have a
fixed memory limit, but the resource needs of the tool containers can fluctuate
depending on the developer's activities.

The proposed Pod Level Resource Spec feature addresses this by allowing all
containers within the pod to share a defined resource pool. In the example
provided, even though the IDE has its own resource constraints, tool1 and tool2
containers can dynamically utilize any remaining resources within the pod's
overall limit. 

```yaml
apiVersion: v1
kind: Pod
metadata:
  labels:
    run: ide
  name: myide
spec:
  resources:
    limits:
      memory: "1024Mi"
      cpu: "4"
  initContainers:
    - image: debian:buster
      name: shell

>       restartPolicy=Always

Reviewer: ```suggestion
      restartPolicy:Always
```

    - image: first-tool:latest
      name: tool1

>       restartPolicy=Always

Reviewer: ```suggestion
      restartPolicy:Always
```

    - image: second-tool:latest
      name: tool2

>       restartPolicy=Always

Reviewer: ```suggestion
      restartPolicy:Always
```

  containers:
    - name: ide
      image: theia:latest
      resources:
        requests:
          memory: "128Mi"
          cpu: "0.5"
        limits:
          memory: "256Mi"
          cpu: "1"
```

Without this feature, the developers will need to estimate beforehand how much
resources to allocate to the tools, a task that can be challenging especially for
workloads with varying resource demands.

> #### Better Resource Utilization

Reviewer: It will be great to have an explainer either in the KEP or later in the blog post clearly articulating how this is different from the burstable pods where contianers can share resources from other containers across the whole node, not just from this container.

others remain underutilized, resulting in inefficiencies.

Moreover, manually tuning requests and limits for each container can be
time-consuming and complex. By defining pod-level limits, all containers can
dynamically share the designated resources. For example, when one container is
performing a resource-intensive task, it can leverage the unused resources of
other containers within the pod. This approach not only improves overall resource
utilization but can also lead to cost savings in cloud environments where
resource allocation impacts billing.

### Notes/Constraints/Caveats

1. cgroupv1 has been moved to maintenance mode since Kubernetes version 1.31.
   Hence this feature will be supported only on cgroupv2. Kubelet will fail to
   admit pods with pod-level resources on nodes with cgroupv1.

2. Pod Level Resource Specifications is an **opt-in** feature, and will have no
effect on existing deployments. Only deployments that explicitly require this
functionality should turn it on by setting the relevant resource section at
pod-level in the Pod specification.

3. In alpha stage of Pod-level Resources feature, when using pod-level resources,

>    in-place pod resizing is unsupported with in-place pod resizing. **Users should

Author: Adding in validation rules in follow up PR.

existing tests to make this code solid enough prior to committing the changes
necessary to implement this enhancement.

#### Unit tests

This feature will touch multiple components. For alpha, unit tests coverage for following packages needs to be added:

* Scheduler logic will be updated to consider pod-level requests. Hence pkg/scheduler will require additional coverage.
* pkg/kubelet/cm will be updated to set pod-level cgroups using CPU requests and limits, and memory limits.
* pkg/apis/core/validation/types_test.go and pkg/apis/core/validation since new fields are added in PodSpec and also new validation rules are required for the new fields.
* pkg/kubeapiserver/admission for changes made in the admission logic for LimitRanger and ResourceQuota admission controllers.

#### e2e tests

Following scenarios need to be covered:

* Cgroup settings when pod-level resources are set.
* Validate scheduling and admission.
* Validate the containers with no limits set are throttled on CPU when CPU usage reaches Pod level CPU limits.
* Validate the containers with no limits set are OOMKilled when memory usage reaches Pod level memory limits.

> 

Reviewer: Test the correct values in `TotalResourcesRequested`

Author: Adding it, although TotalResourcesRequested will be added in beta.

### Graduation Criteria

> #### Phase 1: Alpha (target 1.31)

Reviewer: 1.32?

* Feature is disabled by default. It is an opt-in feature which can be enabled by enabling the `PodLevelResources`
feature gate and by setting the new `resources` fields in PodSpec at Pod level.
* Support the basic functionality for scheduler to consider pod-level resource requests to find a suitable node.
* Support the basic functionality for kubelet to translate pod-level requests/limits to pod-level cgroup settings.
* Unit test coverage.
* E2E tests.
* Documentation mentioning high level design.

> #### Phase 2:  Beta (target 1.32)

Reviewer: 1.33?

> #### Phase 2:  Beta (target 1.32)

Reviewer: so things like cpu manager, limit ranger, etc. are not part of beta graduation criteria? Or those are in alpha?

Author: Support for all features that are scoped out of alpha phase in the KEP design includes CPU manager. 

There's a section for Limit Ranger in alpha.

> #### Phase 2:  Beta (target 1.32)

Reviewer: Is InPlace update support part of beta or not in scope at all?

Author: This is TBD. Added as a bullet point (in a follow up PR that I am working on) though to not forget.
